<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Transform genomic ranges into "absolute" ranges — absoluteRanges • GenomicRanges</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Transform genomic ranges into " absolute ranges absoluteranges><meta property="og:description" content="absoluteRanges transforms the genomic ranges in x into
  absolute ranges i.e. into ranges counted from the beginning of
  the virtual sequence obtained by concatenating all the sequences in the
  underlying genome (in the order reported by seqlevels(x)).
relativeRanges performs the reverse transformation.
NOTE: These functions only work on small genomes. See Details
  section below."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GenomicRanges</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.51.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/GenomicRangesIntroduction.html">An Introduction to the GenomicRanges Package</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/Bioconductor/GenomicRanges/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Transform genomic ranges into "absolute" ranges</h1>
    
    <div class="hidden name"><code>absoluteRanges.Rd</code></div>
    </div>

    <div class="ref-description">
    <p><code>absoluteRanges</code> transforms the genomic ranges in <code>x</code> into
  <em>absolute</em> ranges i.e. into ranges counted from the beginning of
  the virtual sequence obtained by concatenating all the sequences in the
  underlying genome (in the order reported by <code>seqlevels(x)</code>).</p>
<p><code>relativeRanges</code> performs the reverse transformation.</p>
<p>NOTE: These functions only work on <em>small</em> genomes. See Details
  section below.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">absoluteRanges</span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span><span class="fu">relativeRanges</span><span class="op">(</span><span class="va">x</span>, <span class="va">seqlengths</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Related utility:</span></span>
<span><span class="fu">isSmallGenome</span><span class="op">(</span><span class="va">seqlengths</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>x</dt>
<dd><p>For <code>absoluteRanges</code>: a <a href="GRanges-class.html">GenomicRanges</a> object with ranges
    defined on a <em>small</em> genome (see Details section below).</p>
<p>For <code>relativeRanges</code>: an <a href="https://rdrr.io/pkg/IRanges/man/IntegerRanges-class.html" class="external-link">IntegerRanges</a> object.</p></dd>

  <dt>seqlengths</dt>
<dd><p>An object holding sequence lengths. This can be a named integer
    (or numeric) vector with no duplicated names as returned by
    <code><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlengths</a>()</code>, or any object from
    which sequence lengths can be extracted with
    <code><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlengths</a>()</code>.</p>
<p>For <code>relativeRanges</code>, <code>seqlengths</code> must describe a <em>small</em>
    genome (see Details section below).</p></dd>

</dl></div>
    <div id="details">
    <h2>Details</h2>
    <p>Because <code>absoluteRanges</code> returns the <em>absolute</em> ranges in an
  <a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a> object, and because an <a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a>
  object cannot hold ranges with an end &gt; <code>.Machine$integer.max</code>
  (i.e. &gt;= 2^31 on most platforms), <code>absoluteRanges</code> cannot be used
  if the size of the underlying genome (i.e. the total length of the
  sequences in it) is &gt; <code>.Machine$integer.max</code>. Utility function
  <code>isSmallGenome</code> is provided as a mean for the user to check
  upfront whether the genome is <em>small</em> (i.e. its size is &lt;=
  <code>.Machine$integer.max</code>) or not, and thus compatible with
  <code>absoluteRanges</code> or not.</p>
<p><code>relativeRanges</code> applies the same restriction by looking at the
  <code>seqlengths</code> argument.</p>
    </div>
    <div id="value">
    <h2>Value</h2>
    

<p>An <a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a> object for <code>absoluteRanges</code>.</p>


<p>A <a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a> object for <code>relativeRanges</code>.</p>
<p></p>
<p><code>absoluteRanges</code> and <code>relativeRanges</code> both return an object that
  is <em>parallel</em> to the input object (i.e. same length and names).</p>
<p></p>
<p><code>isSmallGenome</code> returns TRUE if the total length of the underlying
  sequences is &lt;= <code>.Machine$integer.max</code> (e.g. Fly genome),
  FALSE if not (e.g. Human genome), or NA if it cannot be computed (because
  some sequence lengths are NA).</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>H. Pagès</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p></p><ul><li><p><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a> objects.</p></li>
<li><p><a href="https://rdrr.io/pkg/IRanges/man/IntegerRanges-class.html" class="external-link">IntegerRanges</a> objects in the <span class="pkg">IRanges</span> package.</p></li>
<li><p><a href="https://rdrr.io/pkg/GenomeInfoDb/man/Seqinfo-class.html" class="external-link">Seqinfo</a> objects and the <code><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlengths</a></code> getter in
          the <span class="pkg">GenomeInfoDb</span> package.</p></li>
<li><p><a href="genomicvars.html">genomicvars</a> for manipulating genomic variables.</p></li>
<li><p>The <code><a href="tileGenome.html">tileGenome</a></code> function for putting tiles on a
          genome.</p></li>
</ul></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">## ---------------------------------------------------------------------</span></span></span>
<span class="r-in"><span><span class="co">## TOY EXAMPLE</span></span></span>
<span class="r-in"><span><span class="co">## ---------------------------------------------------------------------</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Rle-class.html" class="external-link">Rle</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chr2"</span>, <span class="st">"chr1"</span>, <span class="st">"chr3"</span>, <span class="st">"chr1"</span><span class="op">)</span>, <span class="fl">4</span><span class="op">:</span><span class="fl">1</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>              <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, width<span class="op">=</span><span class="fl">5</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>              seqinfo<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/Seqinfo-class.html" class="external-link">Seqinfo</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chr1"</span>, <span class="st">"chr2"</span>, <span class="st">"chr3"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">100</span>, <span class="fl">50</span>, <span class="fl">20</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">ar</span> <span class="op">&lt;-</span> <span class="fu">absoluteRanges</span><span class="op">(</span><span class="va">gr</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">ar</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> IRanges object with 10 ranges and 0 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>            start       end     width</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    [1]       101       105         5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    [2]       102       106         5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    [3]       103       107         5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    [4]       104       108         5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    [5]         5         9         5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    [6]         6        10         5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    [7]         7        11         5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    [8]       158       162         5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    [9]       159       163         5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [10]        10        14         5</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">gr2</span> <span class="op">&lt;-</span> <span class="fu">relativeRanges</span><span class="op">(</span><span class="va">ar</span>, <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlengths</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">gr2</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GRanges object with 10 ranges and 0 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        seqnames    ranges strand</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    [1]     chr2       1-5      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    [2]     chr2       2-6      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    [3]     chr2       3-7      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    [4]     chr2       4-8      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    [5]     chr1       5-9      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    [6]     chr1      6-10      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    [7]     chr1      7-11      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    [8]     chr3      8-12      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    [9]     chr3      9-13      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [10]     chr1     10-14      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 3 sequences from an unspecified genome</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Sanity check:</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/stopifnot.html" class="external-link">stopifnot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="va">gr</span> <span class="op">==</span> <span class="va">gr2</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## ---------------------------------------------------------------------</span></span></span>
<span class="r-in"><span><span class="co">## ON REAL DATA</span></span></span>
<span class="r-in"><span><span class="co">## ---------------------------------------------------------------------</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## With a "small" genome</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">TxDb.Dmelanogaster.UCSC.dm3.ensGene</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">txdb</span> <span class="op">&lt;-</span> <span class="va">TxDb.Dmelanogaster.UCSC.dm3.ensGene</span></span></span>
<span class="r-in"><span><span class="va">ex</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/transcripts.html" class="external-link">exons</a></span><span class="op">(</span><span class="va">txdb</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">ex</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GRanges object with 76920 ranges and 1 metadata column:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>            seqnames        ranges strand |   exon_id</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>               &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       [1]     chr2L     7529-8116      + |         1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       [2]     chr2L     8193-8589      + |         2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       [3]     chr2L     8193-9484      + |         3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       [4]     chr2L     8229-9484      + |         4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       [5]     chr2L     8668-9484      + |         5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       ...       ...           ...    ... .       ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [76916]   chrYHet 327915-328204      - |     76916</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [76917]   chrYHet 328270-328489      - |     76917</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [76918] chrUextra 523024-523048      - |     76918</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [76919] chrUextra 523024-523086      - |     76919</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [76920] chrUextra 523060-523086      - |     76920</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 15 sequences (1 circular) from dm3 genome</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">isSmallGenome</span><span class="op">(</span><span class="va">ex</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] TRUE</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Note that because isSmallGenome() can return NA (see Value section</span></span></span>
<span class="r-in"><span><span class="co">## above), its result should always be wrapped inside isTRUE() when</span></span></span>
<span class="r-in"><span><span class="co">## used in an if statement:</span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Logic.html" class="external-link">isTRUE</a></span><span class="op">(</span><span class="fu">isSmallGenome</span><span class="op">(</span><span class="va">ex</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span>    <span class="va">ar</span> <span class="op">&lt;-</span> <span class="fu">absoluteRanges</span><span class="op">(</span><span class="va">ex</span><span class="op">)</span></span></span>
<span class="r-in"><span>    <span class="va">ar</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>    <span class="va">ex2</span> <span class="op">&lt;-</span> <span class="fu">relativeRanges</span><span class="op">(</span><span class="va">ar</span>, <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlengths</a></span><span class="op">(</span><span class="va">ex</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span>    <span class="va">ex2</span>  <span class="co"># original strand is not restored</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>    <span class="co">## Sanity check:</span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/strand.html" class="external-link">strand</a></span><span class="op">(</span><span class="va">ex2</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/strand.html" class="external-link">strand</a></span><span class="op">(</span><span class="va">ex</span><span class="op">)</span>  <span class="co"># restore the strand</span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://rdrr.io/r/base/stopifnot.html" class="external-link">stopifnot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="va">ex</span> <span class="op">==</span> <span class="va">ex2</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## With a "big" genome (but we can reduce it)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">TxDb.Hsapiens.UCSC.hg19.knownGene</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">txdb</span> <span class="op">&lt;-</span> <span class="va">TxDb.Hsapiens.UCSC.hg19.knownGene</span></span></span>
<span class="r-in"><span><span class="va">ex</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/transcripts.html" class="external-link">exons</a></span><span class="op">(</span><span class="va">txdb</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">isSmallGenome</span><span class="op">(</span><span class="va">ex</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] FALSE</span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span>    <span class="fu">absoluteRanges</span><span class="op">(</span><span class="va">ex</span><span class="op">)</span>  <span class="co"># error!</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## However, if we are only interested in some chromosomes, we might</span></span></span>
<span class="r-in"><span><span class="co">## still be able to use absoluteRanges():</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlevels</a></span><span class="op">(</span><span class="va">ex</span>, pruning.mode<span class="op">=</span><span class="st">"coarse"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">isSmallGenome</span><span class="op">(</span><span class="va">ex</span><span class="op">)</span>  <span class="co"># TRUE!</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] TRUE</span>
<span class="r-in"><span><span class="va">ar</span> <span class="op">&lt;-</span> <span class="fu">absoluteRanges</span><span class="op">(</span><span class="va">ex</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">ex2</span> <span class="op">&lt;-</span> <span class="fu">relativeRanges</span><span class="op">(</span><span class="va">ar</span>, <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlengths</a></span><span class="op">(</span><span class="va">ex</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Sanity check:</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/strand.html" class="external-link">strand</a></span><span class="op">(</span><span class="va">ex2</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/strand.html" class="external-link">strand</a></span><span class="op">(</span><span class="va">ex</span><span class="op">)</span> </span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/stopifnot.html" class="external-link">stopifnot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="va">ex</span> <span class="op">==</span> <span class="va">ex2</span><span class="op">)</span><span class="op">)</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Patrick Aboyoun, Hervé Pagès, Michael Lawrence.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

