<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Memory-efficient representation of genomic positions — GPos-class • GenomicRanges</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Memory-efficient representation of genomic positions — GPos-class"><meta property="og:description" content="The GPos class is a container for storing a set of genomic positions
  (a.k.a. genomic loci).
  It exists in 2 flavors: UnstitchedGPos and StitchedGPos. Each flavor uses
  a particular internal representation:
In an UnstitchedGPos instance the positions are stored as an
          integer vector.
In a StitchedGPos instance the positions are stored as an
          IRanges object where each range represents a run of
          consecutive positions (i.e. a run of positions that
          are adjacent and in ascending order). This storage is
          particularly memory-efficient when the vector of positions
          contains long runs of consecutive positions.
Because genomic positions can be seen as genomic ranges of width 1, the
  GPos class extends the GenomicRanges virtual class (via the
  GRanges class)."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GenomicRanges</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.51.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/GenomicRangesIntroduction.html">An Introduction to the GenomicRanges Package</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/Bioconductor/GenomicRanges/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Memory-efficient representation of genomic positions</h1>
    
    <div class="hidden name"><code>GPos-class.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>The GPos class is a container for storing a set of <em>genomic positions</em>
  (a.k.a. <em>genomic loci</em>).
  It exists in 2 flavors: UnstitchedGPos and StitchedGPos. Each flavor uses
  a particular internal representation:</p><ul><li><p>In an UnstitchedGPos instance the positions are stored as an
          integer vector.</p></li>
<li><p>In a StitchedGPos instance the positions are stored as an
          IRanges object where each range represents a run of
          <em>consecutive positions</em> (i.e. a run of positions that
          are adjacent and in <em>ascending order</em>). This storage is
          particularly memory-efficient when the vector of positions
          contains long runs of consecutive positions.</p></li>
</ul><p>Because genomic positions can be seen as genomic ranges of width 1, the
  GPos class extends the <a href="GRanges-class.html">GenomicRanges</a> virtual class (via the
  GRanges class).</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="co">## Constructor function</span></span>
<span><span class="fu">GPos</span><span class="op">(</span>seqnames<span class="op">=</span><span class="cn">NULL</span>, pos<span class="op">=</span><span class="cn">NULL</span>, strand<span class="op">=</span><span class="cn">NULL</span>,</span>
<span>     <span class="va">...</span>, seqinfo<span class="op">=</span><span class="cn">NULL</span>, seqlengths<span class="op">=</span><span class="cn">NULL</span>, stitch<span class="op">=</span><span class="cn">NA</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>seqnames, strand, ..., seqinfo, seqlengths</dt>
<dd><p>See documentation of the <code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a>()</code> constructor function
    for a description of these arguments.</p></dd>

  <dt>pos</dt>
<dd><p><code>NULL</code>, or an integer or numeric vector, or an <a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a>
    or <a href="https://rdrr.io/pkg/IRanges/man/IPos-class.html" class="external-link">IPos</a> object (or other <a href="https://rdrr.io/pkg/IRanges/man/IntegerRanges-class.html" class="external-link">IntegerRanges</a>
    derivative).
    If not <code>NULL</code>, <code>GPos()</code> will try to turn it into an
    <a href="https://rdrr.io/pkg/IRanges/man/IPos-class.html" class="external-link">IPos</a> derivative with <code>IPos(pos, stitch=stitch)</code>.</p>
<p>When <code>pos</code> is an <a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a> object (or other <a href="https://rdrr.io/pkg/IRanges/man/IntegerRanges-class.html" class="external-link">IntegerRanges</a>
    derivative), each range in it is interpreted as a run of
    consecutive positions.</p></dd>

  <dt>stitch</dt>
<dd><p><code>TRUE</code>, <code>FALSE</code>, or <code>NA</code> (the default).</p>
<p>Controls which internal representation should be used: StitchedGPos
    (when <code>stitch</code> is <code>TRUE</code>) or UnstitchedGPos (when <code>stitch</code>
    is <code>FALSE</code>).</p>
<p>When <code>stitch</code> is <code>NA</code> (the default), which internal
    representation will be used depends on the flavour of the
    <a href="https://rdrr.io/pkg/IRanges/man/IPos-class.html" class="external-link">IPos</a> derivative returned by <code>IPos(pos)</code>:
    UnstitchedGPos if <code>IPos(pos)</code> returns an
    <a href="https://rdrr.io/pkg/IRanges/man/IPos-class.html" class="external-link">UnstitchedIPos</a> instance, and StitchedGPos if it
    returns a <a href="https://rdrr.io/pkg/IRanges/man/IPos-class.html" class="external-link">StitchedIPos</a> instance.</p></dd>

</dl></div>
    <div id="details">
    <h2>Details</h2>
    <p>Even though a <a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a> object can be used for storing genomic
  positions, using a GPos object is more efficient. In particular the
  memory footprint of an UnstitchedGPos object is typically about half
  that of a <a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a> object.</p>
<p>OTOH the memory footprint of a StitchedGPos object can vary a lot but
  will never be worse than that of a <a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a> object. However it
  will reduce dramatically if the vector of positions contains long runs
  of consecutive positions. In the worst case scenario (i.e. when the
  object contains no consecutive positions) its memory footprint will
  be the same as that of a <a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a> object.</p>
<p>Like for any <a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">Vector</a> derivative, the length of a
  GPos object cannot exceed <code>.Machine$integer.max</code> (i.e. 2^31 on
  most platforms). <code>GPos()</code> will return an error if <code>pos</code>
  contains too many positions.</p>
    </div>
    <div id="value">
    <h2>Value</h2>
    

<p>An UnstitchedGPos or StitchedGPos object.</p>
    </div>
    <div id="accessors">
    <h2>Accessors</h2>
    

<p></p><div class="section">
<h3 id="getters">Getters<a class="anchor" aria-label="anchor" href="#getters"></a></h3>
<p>GPos objects support the same set of getters as other <a href="GRanges-class.html">GenomicRanges</a>
    derivatives (i.e. <code><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqnames()</a></code>, <code><a href="https://rdrr.io/pkg/IRanges/man/range-squeezers.html" class="external-link">ranges()</a></code>, <code><a href="https://rdrr.io/r/stats/start.html" class="external-link">start()</a></code>,
    <code><a href="https://rdrr.io/r/stats/start.html" class="external-link">end()</a></code>, <code><a href="https://rdrr.io/pkg/BiocGenerics/man/strand.html" class="external-link">strand()</a></code>, <code><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">mcols()</a></code>, <code><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqinfo()</a></code>,
    etc...), plus the <code><a href="https://rdrr.io/pkg/BiocGenerics/man/start.html" class="external-link">pos()</a></code> getter which is equivalent to
    <code><a href="https://rdrr.io/r/stats/start.html" class="external-link">start()</a></code> or <code><a href="https://rdrr.io/r/stats/start.html" class="external-link">end()</a></code>. See <code>?<a href="GRanges-class.html">GenomicRanges</a></code> for the
    list of getters supported by <a href="GRanges-class.html">GenomicRanges</a> derivatives.</p>
<p>Note that <code><a href="https://rdrr.io/pkg/IRanges/man/range-squeezers.html" class="external-link">ranges()</a></code> returns an <a href="https://rdrr.io/pkg/IRanges/man/IPos-class.html" class="external-link">IPos</a> derivative
    instead of the <a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a> object that one gets with other
    <a href="GRanges-class.html">GenomicRanges</a> derivatives. To get an <a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a>
    object, you need to call <code><a href="https://rdrr.io/pkg/IRanges/man/range-squeezers.html" class="external-link">ranges()</a></code> again on this
    <a href="https://rdrr.io/pkg/IRanges/man/IPos-class.html" class="external-link">IPos</a> derivative i.e. do <code>ranges(ranges(x))</code>
    on GPos object <code>x</code>.</p>
</div>

<p></p><div class="section">
<h3 id="setters">Setters<a class="anchor" aria-label="anchor" href="#setters"></a></h3>
<p>Like <a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a> objects, GPos derivatives support the following
    setters:</p><ul><li><p>The <code><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqnames()</a></code> and <code><a href="https://rdrr.io/pkg/BiocGenerics/man/strand.html" class="external-link">strand()</a></code> setters.</p></li>
<li><p>The <code><a href="https://rdrr.io/r/base/names.html" class="external-link">names()</a></code>, <code><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">mcols()</a></code> and <code><a href="https://rdrr.io/pkg/S4Vectors/man/Annotated-class.html" class="external-link">metadata()</a></code> setters.</p></li>
<li><p>The family of setters that operate on the <code>seqinfo</code>
            component of an object:
            <code><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlevels</a>()</code>,
            <code><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqlevelsStyle.html" class="external-link">seqlevelsStyle</a>()</code>,
            <code><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlengths</a>()</code>,
            <code><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">isCircular</a>()</code>,
            <code><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">genome</a>()</code>,
            and <code><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqinfo</a>()</code>.
            These setters are defined and documented in the <span class="pkg">GenomeInfoDb</span>
            package.</p></li>
</ul><p>However, there is no <code><a href="https://rdrr.io/pkg/BiocGenerics/man/start.html" class="external-link">pos()</a></code> setter for GPos derivatives at the
    moment (although one might be added in the future).</p>
</div>

    </div>
    <div id="coercion">
    <h2>Coercion</h2>
    <p>From UnstitchedGPos to StitchedGPos and vice-versa: coercion back and
  forth between UnstitchedGPos and StitchedGPos is supported via
  <code>as(x, "StitchedGPos")</code> and <code>as(x, "UnstitchedGPos")</code>.
  This is the most efficient and recommended way to switch between the
  2 internal representations. Note that this switch can have dramatic
  consequences on memory usage so is for advanced users only.
  End users should almost never need to do this switch when following
  a typical workflow.</p>
<p>From <a href="GRanges-class.html">GenomicRanges</a> to UnstitchedGPos, StitchedGPos, or GPos:
  A <a href="GRanges-class.html">GenomicRanges</a> derivative <code>x</code> in which all the ranges have
  a width of 1 can be coerced to an UnstitchedGPos or StitchedGPos object
  with <code>as(x, "UnstitchedGPos")</code> or <code>as(x, "StitchedGPos")</code>,
  respectively.
  For convenience <code>as(x, "GPos")</code> is supported and is equivalent to
  <code>as(x, "UnstitchedGPos")</code>.</p>
<p>From GPos to <a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a>:
  A GPos derivative <code>x</code> can be coerced to a <a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a> object
  with <code>as(x, "GRanges")</code>. However be aware that the resulting object
  can use thousands times (or more) memory than <code>x</code>!
  See "MEMORY USAGE" in the Examples section below.</p>
<p>From GPos to ordinary R objects:
  Like with any other <a href="GRanges-class.html">GenomicRanges</a> derivative, <code><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character()</a></code>,
  <code><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor()</a></code>, and <code><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame()</a></code> work on a GPos derivative
  <code>x</code>. Note however that <code>as.data.frame(x)</code> returns a data frame
  with a <code>pos</code> column (containing <code>pos(x)</code>) instead of the
  <code>start</code>, <code>end</code>, and <code>width</code> columns that one gets with other
  <a href="GRanges-class.html">GenomicRanges</a> derivatives.</p>
    </div>
    <div id="subsetting">
    <h2>Subsetting</h2>
    <p>A GPos derivative can be subsetted exactly like a <a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a> object.</p>
    </div>
    <div id="concatenation">
    <h2>Concatenation</h2>
    <p>GPos derivatives can be concatenated with <code><a href="https://rdrr.io/r/base/c.html" class="external-link">c()</a></code> or <code><a href="https://rdrr.io/pkg/BiocGenerics/man/append.html" class="external-link">append()</a></code>.
  See <code>?<a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></code> in the <span class="pkg">S4Vectors</span> package for
  more information about concatenating Vector derivatives.</p>
    </div>
    <div id="splitting-and-relisting">
    <h2>Splitting and Relisting</h2>
    <p>Like with any other <a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a> object, <code><a href="https://rdrr.io/pkg/S4Vectors/man/splitAsList.html" class="external-link">split()</a></code> and <code><a href="https://rdrr.io/r/utils/relist.html" class="external-link">relist()</a></code>
  work on a GPos derivative.</p>
    </div>
    <div id="note">
    <h2>Note</h2>
    <p>Internal representation of GPos objects has changed in <span class="pkg">GenomicRanges</span>
  1.29.10 (Bioc 3.6). Update any old object <code>x</code> with:
  <code>x &lt;- updateObject(x, verbose=TRUE)</code>.</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Hervé Pagès; based on ideas borrowed from Georg Stricker
  <a href="mailto:georg.stricker@in.tum.de">georg.stricker@in.tum.de</a> and Julien Gagneur
  <a href="mailto:gagneur@in.tum.de">gagneur@in.tum.de</a></p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p></p><ul><li><p>The <a href="https://rdrr.io/pkg/IRanges/man/IPos-class.html" class="external-link">IPos</a> class in the <span class="pkg">IRanges</span>
          package for storing a set of <em>integer positions</em> (i.e.
          integer ranges of width 1).</p></li>
<li><p>The <a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a> class for storing a set of <em>genomic ranges</em>
          of arbitrary width.</p></li>
<li><p><a href="https://rdrr.io/pkg/GenomeInfoDb/man/Seqinfo-class.html" class="external-link">Seqinfo</a> objects and the
          <code><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqinfo</a></code> accessor and family in the
          <span class="pkg">GenomeInfoDb</span> package for accessing/modifying information
          about the underlying sequences of a <a href="GRanges-class.html">GenomicRanges</a> derivative.</p></li>
<li><p><a href="GenomicRanges-comparison.html">GenomicRanges-comparison</a> for comparing and ordering genomic
          ranges and/or positions.</p></li>
<li><p><a href="findOverlaps-methods.html">findOverlaps-methods</a> for finding overlapping
          genomic ranges and/or positions.</p></li>
<li><p><a href="intra-range-methods.html">intra-range-methods</a> and
          <a href="inter-range-methods.html">inter-range-methods</a>
          for <em>intra range</em> and <em>inter range</em> transformations
          of <a href="GRanges-class.html">GenomicRanges</a> derivatives.</p></li>
<li><p><a href="coverage-methods.html">coverage-methods</a> for computing the coverage
          of a set of genomic ranges and/or positions.</p></li>
<li><p><a href="nearest-methods.html">nearest-methods</a> for finding the nearest
          genomic range/position neighbor.</p></li>
<li><p>The <code><a href="https://rdrr.io/pkg/BSgenome/man/SNPlocs-class.html" class="external-link">snpsBySeqname</a></code>,
          <code><a href="https://rdrr.io/pkg/BSgenome/man/SNPlocs-class.html" class="external-link">snpsByOverlaps</a></code>, and
          <code><a href="https://rdrr.io/pkg/BSgenome/man/SNPlocs-class.html" class="external-link">snpsById</a></code> methods for
          <a href="https://rdrr.io/pkg/BSgenome/man/SNPlocs-class.html" class="external-link">SNPlocs</a> objects defined in the <span class="pkg">BSgenome</span>
          package for extractors that return a GPos derivative.</p></li>
<li><p><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html" class="external-link">SummarizedExperiment</a> objects and
          derivatives in the <span class="pkg">SummarizedExperiment</span> package.</p></li>
</ul></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu">showClass</span><span class="op">(</span><span class="st">"GPos"</span><span class="op">)</span>  <span class="co"># shows the known subclasses</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Virtual Class "GPos" [package "GenomicRanges"]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Slots:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                                                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Name:           ranges elementMetadata     elementType        metadata</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Class:            IPos       DataFrame       character            list</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Name:         seqnames          strand         seqinfo</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Class:             Rle             Rle         Seqinfo</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Extends: </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Class "GenomicPos", directly</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Class "GRanges", directly</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Class "GenomicRanges", by class "GenomicPos", distance 2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Class "Pos", by class "GenomicPos", distance 2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Class "Ranges", by class "GenomicPos", distance 3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Class "GenomicRanges_OR_missing", by class "GenomicPos", distance 3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Class "List", by class "GenomicPos", distance 4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Class "Vector", by class "GenomicPos", distance 5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Class "list_OR_List", by class "GenomicPos", distance 5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Class "Annotated", by class "GenomicPos", distance 6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Class "vector_OR_Vector", by class "GenomicPos", distance 6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Class "GenomicRanges_OR_GenomicRangesList", by class "GenomicRanges", distance 3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Class "GenomicRanges_OR_GRangesList", by class "GenomicRanges", distance 3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Known Subclasses: "UnstitchedGPos", "StitchedGPos"</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## ---------------------------------------------------------------------</span></span></span>
<span class="r-in"><span><span class="co">## BASIC EXAMPLES</span></span></span>
<span class="r-in"><span><span class="co">## ---------------------------------------------------------------------</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Example 1:</span></span></span>
<span class="r-in"><span><span class="va">gpos1a</span> <span class="op">&lt;-</span> <span class="fu">GPos</span><span class="op">(</span>seqnames<span class="op">=</span><span class="fu">Rle</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chr1"</span>, <span class="st">"chr2"</span>, <span class="st">"chr1"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">6</span>, <span class="fl">4</span><span class="op">)</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>               pos<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">44</span><span class="op">:</span><span class="fl">53</span>, <span class="fl">5</span><span class="op">:</span><span class="fl">10</span>, <span class="fl">2</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">gpos1a</span>  <span class="co"># unstitched</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> UnstitchedGPos object with 20 positions and 0 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        seqnames       pos strand</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           &lt;Rle&gt; &lt;integer&gt;  &lt;Rle&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    [1]     chr1        44      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    [2]     chr1        45      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    [3]     chr1        46      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    [4]     chr1        47      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    [5]     chr1        48      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    ...      ...       ...    ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [16]     chr2        10      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [17]     chr1         2      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [18]     chr1         3      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [19]     chr1         4      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [20]     chr1         5      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">gpos1a</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 20</span>
<span class="r-in"><span><span class="fu">seqnames</span><span class="op">(</span><span class="va">gpos1a</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> factor-Rle of length 20 with 3 runs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Lengths:   10    6    4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Values : chr1 chr2 chr1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Levels(2): chr1 chr2</span>
<span class="r-in"><span><span class="fu">pos</span><span class="op">(</span><span class="va">gpos1a</span><span class="op">)</span>  <span class="co"># same as 'start(gpos1a)' and 'end(gpos1a)'</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1] 44 45 46 47 48 49 50 51 52 53  5  6  7  8  9 10  2  3  4  5</span>
<span class="r-in"><span><span class="fu">strand</span><span class="op">(</span><span class="va">gpos1a</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> factor-Rle of length 20 with 1 run</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Lengths: 20</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Values :  *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Levels(3): + - *</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">gpos1a</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1] "chr1:44" "chr1:45" "chr1:46" "chr1:47" "chr1:48" "chr1:49" "chr1:50"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [8] "chr1:51" "chr1:52" "chr1:53" "chr2:5"  "chr2:6"  "chr2:7"  "chr2:8" </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [15] "chr2:9"  "chr2:10" "chr1:2"  "chr1:3"  "chr1:4"  "chr1:5" </span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">gpos1a</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    seqnames pos strand</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1      chr1  44      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2      chr1  45      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3      chr1  46      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4      chr1  47      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5      chr1  48      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6      chr1  49      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7      chr1  50      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8      chr1  51      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 9      chr1  52      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10     chr1  53      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 11     chr2   5      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 12     chr2   6      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 13     chr2   7      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 14     chr2   8      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 15     chr2   9      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 16     chr2  10      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 17     chr1   2      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 18     chr1   3      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 19     chr1   4      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 20     chr1   5      *</span>
<span class="r-in"><span><span class="fu">as</span><span class="op">(</span><span class="va">gpos1a</span>, <span class="st">"GRanges"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GRanges object with 20 ranges and 0 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        seqnames    ranges strand</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    [1]     chr1        44      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    [2]     chr1        45      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    [3]     chr1        46      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    [4]     chr1        47      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    [5]     chr1        48      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    ...      ...       ...    ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [16]     chr2        10      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [17]     chr1         2      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [18]     chr1         3      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [19]     chr1         4      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [20]     chr1         5      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu">as</span><span class="op">(</span><span class="va">gpos1a</span>, <span class="st">"GRanges"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    seqnames start end width strand</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1      chr1    44  44     1      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2      chr1    45  45     1      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3      chr1    46  46     1      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4      chr1    47  47     1      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5      chr1    48  48     1      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6      chr1    49  49     1      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7      chr1    50  50     1      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8      chr1    51  51     1      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 9      chr1    52  52     1      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10     chr1    53  53     1      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 11     chr2     5   5     1      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 12     chr2     6   6     1      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 13     chr2     7   7     1      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 14     chr2     8   8     1      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 15     chr2     9   9     1      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 16     chr2    10  10     1      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 17     chr1     2   2     1      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 18     chr1     3   3     1      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 19     chr1     4   4     1      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 20     chr1     5   5     1      *</span>
<span class="r-in"><span><span class="va">gpos1a</span><span class="op">[</span><span class="fl">9</span><span class="op">:</span><span class="fl">17</span><span class="op">]</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> UnstitchedGPos object with 9 positions and 0 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       seqnames       pos strand</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          &lt;Rle&gt; &lt;integer&gt;  &lt;Rle&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1]     chr1        52      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [2]     chr1        53      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [3]     chr2         5      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [4]     chr2         6      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [5]     chr2         7      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [6]     chr2         8      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [7]     chr2         9      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [8]     chr2        10      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [9]     chr1         2      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">gpos1b</span> <span class="op">&lt;-</span> <span class="fu">GPos</span><span class="op">(</span>seqnames<span class="op">=</span><span class="fu">Rle</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chr1"</span>, <span class="st">"chr2"</span>, <span class="st">"chr1"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">6</span>, <span class="fl">4</span><span class="op">)</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>               pos<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">44</span><span class="op">:</span><span class="fl">53</span>, <span class="fl">5</span><span class="op">:</span><span class="fl">10</span>, <span class="fl">2</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span>, stitch<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">gpos1b</span>  <span class="co"># stitched</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> StitchedGPos object with 20 positions and 0 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        seqnames       pos strand</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           &lt;Rle&gt; &lt;integer&gt;  &lt;Rle&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    [1]     chr1        44      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    [2]     chr1        45      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    [3]     chr1        46      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    [4]     chr1        47      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    [5]     chr1        48      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    ...      ...       ...    ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [16]     chr2        10      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [17]     chr1         2      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [18]     chr1         3      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [19]     chr1         4      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [20]     chr1         5      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## 'gpos1a' and 'gpos1b' are semantically equivalent, only their</span></span></span>
<span class="r-in"><span><span class="co">## internal representations differ:</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="va">gpos1a</span> <span class="op">==</span> <span class="va">gpos1b</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] TRUE</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">gpos1c</span> <span class="op">&lt;-</span> <span class="fu">GPos</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chr1:44-53"</span>, <span class="st">"chr2:5-10"</span>, <span class="st">"chr1:2-5"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">gpos1c</span>  <span class="co"># stitched</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> StitchedGPos object with 20 positions and 0 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        seqnames       pos strand</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           &lt;Rle&gt; &lt;integer&gt;  &lt;Rle&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    [1]     chr1        44      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    [2]     chr1        45      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    [3]     chr1        46      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    [4]     chr1        47      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    [5]     chr1        48      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    ...      ...       ...    ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [16]     chr2        10      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [17]     chr1         2      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [18]     chr1         3      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [19]     chr1         4      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [20]     chr1         5      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/identical.html" class="external-link">identical</a></span><span class="op">(</span><span class="va">gpos1b</span>, <span class="va">gpos1c</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] TRUE</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Example 2:</span></span></span>
<span class="r-in"><span><span class="va">pos_runs</span> <span class="op">&lt;-</span> <span class="fu">GRanges</span><span class="op">(</span><span class="st">"chrI"</span>, <span class="fu">IRanges</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">6</span>, <span class="fl">12</span>, <span class="fl">17</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">10</span>, <span class="fl">16</span>, <span class="fl">20</span><span class="op">)</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                    strand<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"*"</span>, <span class="st">"-"</span>, <span class="st">"-"</span>, <span class="st">"+"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">gpos2</span> <span class="op">&lt;-</span> <span class="fu">GPos</span><span class="op">(</span><span class="va">pos_runs</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">gpos2</span>  <span class="co"># stitched</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> StitchedGPos object with 19 positions and 0 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        seqnames       pos strand</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           &lt;Rle&gt; &lt;integer&gt;  &lt;Rle&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    [1]     chrI         1      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    [2]     chrI         2      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    [3]     chrI         3      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    [4]     chrI         4      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    [5]     chrI         5      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    ...      ...       ...    ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [15]     chrI        16      -</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [16]     chrI        17      +</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [17]     chrI        18      +</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [18]     chrI        19      +</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [19]     chrI        20      +</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span>
<span class="r-in"><span><span class="fu">strand</span><span class="op">(</span><span class="va">gpos2</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> factor-Rle of length 19 with 3 runs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Lengths:  5 10  4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Values :  *  -  +</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Levels(3): + - *</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Example 3:</span></span></span>
<span class="r-in"><span><span class="va">gpos3A</span> <span class="op">&lt;-</span> <span class="va">gpos3B</span> <span class="op">&lt;-</span> <span class="fu">GPos</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chrI:1-1000"</span>, <span class="st">"chrI:1005-2000"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">npos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">gpos3A</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">mcols</span><span class="op">(</span><span class="va">gpos3A</span><span class="op">)</span><span class="op">$</span><span class="va">sample</span> <span class="op">&lt;-</span> <span class="fu">Rle</span><span class="op">(</span><span class="st">"sA"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">sA_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">10</span>, <span class="va">npos</span>, replace<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">mcols</span><span class="op">(</span><span class="va">gpos3A</span><span class="op">)</span><span class="op">$</span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="va">sA_counts</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">mcols</span><span class="op">(</span><span class="va">gpos3B</span><span class="op">)</span><span class="op">$</span><span class="va">sample</span> <span class="op">&lt;-</span> <span class="fu">Rle</span><span class="op">(</span><span class="st">"sB"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">sB_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">10</span>, <span class="va">npos</span>, replace<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">mcols</span><span class="op">(</span><span class="va">gpos3B</span><span class="op">)</span><span class="op">$</span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="va">sB_counts</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">gpos3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">gpos3A</span>, <span class="va">gpos3B</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">gpos3</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> StitchedGPos object with 3992 positions and 2 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          seqnames       pos strand | sample    counts</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             &lt;Rle&gt; &lt;integer&gt;  &lt;Rle&gt; |  &lt;Rle&gt; &lt;integer&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      [1]     chrI         1      * |     sA         9</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      [2]     chrI         2      * |     sA         9</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      [3]     chrI         3      * |     sA        10</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      [4]     chrI         4      * |     sA        10</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      [5]     chrI         5      * |     sA         7</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      ...      ...       ...    ... .    ...       ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [3988]     chrI      1996      * |     sB         7</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [3989]     chrI      1997      * |     sB         7</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [3990]     chrI      1998      * |     sB         1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [3991]     chrI      1999      * |     sB         6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [3992]     chrI      2000      * |     sB         8</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Example 4:</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">BSgenome.Scerevisiae.UCSC.sacCer2</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: BSgenome</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: Biostrings</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: XVector</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘Biostrings’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following object is masked from ‘package:base’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     strsplit</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: rtracklayer</span>
<span class="r-in"><span><span class="va">genome</span> <span class="op">&lt;-</span> <span class="va">BSgenome.Scerevisiae.UCSC.sacCer2</span></span></span>
<span class="r-in"><span><span class="va">gpos4</span> <span class="op">&lt;-</span> <span class="fu">GPos</span><span class="op">(</span><span class="fu">seqinfo</span><span class="op">(</span><span class="va">genome</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">gpos4</span>  <span class="co"># all the positions along the genome are represented</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> StitchedGPos object with 12162995 positions and 0 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>              seqnames       pos strand</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                 &lt;Rle&gt; &lt;integer&gt;  &lt;Rle&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          [1]     chrI         1      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          [2]     chrI         2      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          [3]     chrI         3      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          [4]     chrI         4      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          [5]     chrI         5      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          ...      ...       ...    ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [12162991]  2micron      6314      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [12162992]  2micron      6315      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [12162993]  2micron      6316      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [12162994]  2micron      6317      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [12162995]  2micron      6318      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 18 sequences (2 circular) from sacCer2 genome</span>
<span class="r-in"><span><span class="fu">mcols</span><span class="op">(</span><span class="va">gpos4</span><span class="op">)</span><span class="op">$</span><span class="va">dna</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="st">"c"</span>, <span class="fu"><a href="https://rdrr.io/r/base/unname.html" class="external-link">unname</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">as.list</a></span><span class="op">(</span><span class="va">genome</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">gpos4</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> StitchedGPos object with 12162995 positions and 1 metadata column:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>              seqnames       pos strand |         dna</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                 &lt;Rle&gt; &lt;integer&gt;  &lt;Rle&gt; | &lt;DNAString&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          [1]     chrI         1      * |           C</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          [2]     chrI         2      * |           C</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          [3]     chrI         3      * |           A</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          [4]     chrI         4      * |           C</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          [5]     chrI         5      * |           A</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          ...      ...       ...    ... .         ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [12162991]  2micron      6314      * |           A</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [12162992]  2micron      6315      * |           A</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [12162993]  2micron      6316      * |           C</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [12162994]  2micron      6317      * |           G</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [12162995]  2micron      6318      * |           A</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 18 sequences (2 circular) from sacCer2 genome</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Note however that, like for any Vector derivative, the length of a</span></span></span>
<span class="r-in"><span><span class="co">## GPos derivative cannot exceed '.Machine$integer.max' (i.e. 2^31 on</span></span></span>
<span class="r-in"><span><span class="co">## most platforms) so the above only works with a "small" genome.</span></span></span>
<span class="r-in"><span><span class="co">## For example it doesn't work with the Human genome:</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">TxDb.Hsapiens.UCSC.hg19.knownGene</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: GenomicFeatures</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: AnnotationDbi</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: Biobase</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Welcome to Bioconductor</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     Vignettes contain introductory material; view with</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     'browseVignettes()'. To cite Bioconductor, see</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     'citation("Biobase")', and for packages 'citation("pkgname")'.</span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span>  <span class="fu">GPos</span><span class="op">(</span><span class="fu">seqinfo</span><span class="op">(</span><span class="va">TxDb.Hsapiens.UCSC.hg19.knownGene</span><span class="op">)</span><span class="op">)</span>  <span class="co"># error!</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## You can use isSmallGenome() to check upfront whether the genome is</span></span></span>
<span class="r-in"><span><span class="co">## "small" or not.</span></span></span>
<span class="r-in"><span><span class="fu"><a href="absoluteRanges.html">isSmallGenome</a></span><span class="op">(</span><span class="va">genome</span><span class="op">)</span>  <span class="co"># TRUE</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] TRUE</span>
<span class="r-in"><span><span class="fu"><a href="absoluteRanges.html">isSmallGenome</a></span><span class="op">(</span><span class="va">TxDb.Hsapiens.UCSC.hg19.knownGene</span><span class="op">)</span>  <span class="co"># FALSE</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] FALSE</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## ---------------------------------------------------------------------</span></span></span>
<span class="r-in"><span><span class="co">## MEMORY USAGE</span></span></span>
<span class="r-in"><span><span class="co">## ---------------------------------------------------------------------</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Coercion to GRanges works...</span></span></span>
<span class="r-in"><span><span class="va">gr4</span> <span class="op">&lt;-</span> <span class="fu">as</span><span class="op">(</span><span class="va">gpos4</span>, <span class="st">"GRanges"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">gr4</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GRanges object with 12162995 ranges and 1 metadata column:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>              seqnames    ranges strand |         dna</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                 &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;DNAString&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          [1]     chrI         1      * |           C</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          [2]     chrI         2      * |           C</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          [3]     chrI         3      * |           A</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          [4]     chrI         4      * |           C</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          [5]     chrI         5      * |           A</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          ...      ...       ...    ... .         ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [12162991]  2micron      6314      * |           A</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [12162992]  2micron      6315      * |           A</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [12162993]  2micron      6316      * |           C</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [12162994]  2micron      6317      * |           G</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [12162995]  2micron      6318      * |           A</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 18 sequences (2 circular) from sacCer2 genome</span>
<span class="r-in"><span><span class="co">## ... but is generally not a good idea:</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/object.size.html" class="external-link">object.size</a></span><span class="op">(</span><span class="va">gpos4</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 12179072 bytes</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/object.size.html" class="external-link">object.size</a></span><span class="op">(</span><span class="va">gr4</span><span class="op">)</span>     <span class="co"># 8 times bigger than the StitchedGPos object!</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 109481200 bytes</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Shuffling the order of the positions impacts memory usage:</span></span></span>
<span class="r-in"><span><span class="va">gpos4r</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rev.html" class="external-link">rev</a></span><span class="op">(</span><span class="va">gpos4</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/object.size.html" class="external-link">object.size</a></span><span class="op">(</span><span class="va">gpos4r</span><span class="op">)</span>  <span class="co"># significantly</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 109482784 bytes</span>
<span class="r-in"><span><span class="va">gpos4s</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">gpos4</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/object.size.html" class="external-link">object.size</a></span><span class="op">(</span><span class="va">gpos4s</span><span class="op">)</span>  <span class="co"># even worse!</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 199548064 bytes</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## If one anticipates a lot of shuffling of the genomic positions,</span></span></span>
<span class="r-in"><span><span class="co">## then an UnstitchedGPos object should be used instead:</span></span></span>
<span class="r-in"><span><span class="va">gpos4b</span> <span class="op">&lt;-</span> <span class="fu">as</span><span class="op">(</span><span class="va">gpos4</span>, <span class="st">"UnstitchedGPos"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/object.size.html" class="external-link">object.size</a></span><span class="op">(</span><span class="va">gpos4b</span><span class="op">)</span>  <span class="co"># initial size is bigger than stitched version</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 60829072 bytes</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/object.size.html" class="external-link">object.size</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rev.html" class="external-link">rev</a></span><span class="op">(</span><span class="va">gpos4b</span><span class="op">)</span><span class="op">)</span>  <span class="co"># size didn't change</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 60829072 bytes</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/object.size.html" class="external-link">object.size</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">gpos4b</span><span class="op">)</span><span class="op">)</span>  <span class="co"># size increased, but is still &lt; stitched</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 150898784 bytes</span>
<span class="r-in"><span>                             <span class="co"># version</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## AN IMPORTANT NOTE: In the worst situations, GPos still performs as</span></span></span>
<span class="r-in"><span><span class="co">## good as a GRanges object.</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/object.size.html" class="external-link">object.size</a></span><span class="op">(</span><span class="fu">as</span><span class="op">(</span><span class="va">gpos4r</span>, <span class="st">"GRanges"</span><span class="op">)</span><span class="op">)</span>  <span class="co"># same size as 'gpos4r'</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 109481200 bytes</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/object.size.html" class="external-link">object.size</a></span><span class="op">(</span><span class="fu">as</span><span class="op">(</span><span class="va">gpos4s</span>, <span class="st">"GRanges"</span><span class="op">)</span><span class="op">)</span>  <span class="co"># same size as 'gpos4s'</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 199546560 bytes</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Best case scenario is when the object is strictly sorted (i.e.</span></span></span>
<span class="r-in"><span><span class="co">## positions are in strict ascending order).</span></span></span>
<span class="r-in"><span><span class="co">## This can be checked with:</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/is.unsorted.html" class="external-link">is.unsorted</a></span><span class="op">(</span><span class="va">gpos4</span>, strict<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>  <span class="co"># 'gpos4' is strictly sorted</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] FALSE</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## ---------------------------------------------------------------------</span></span></span>
<span class="r-in"><span><span class="co">## USING MEMORY-EFFICIENT METADATA COLUMNS</span></span></span>
<span class="r-in"><span><span class="co">## ---------------------------------------------------------------------</span></span></span>
<span class="r-in"><span><span class="co">## In order to keep memory usage as low as possible, it is recommended</span></span></span>
<span class="r-in"><span><span class="co">## to use a memory-efficient representation of the metadata columns that</span></span></span>
<span class="r-in"><span><span class="co">## we want to set on the object. Rle's are particularly well suited for</span></span></span>
<span class="r-in"><span><span class="co">## this, especially if the metadata columns contain long runs of</span></span></span>
<span class="r-in"><span><span class="co">## identical values. This is the case for example if we want to use a</span></span></span>
<span class="r-in"><span><span class="co">## GPos object to represent the coverage of sequencing reads along a</span></span></span>
<span class="r-in"><span><span class="co">## genome.</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Example 5:</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">pasillaBamSubset</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/Rsamtools" class="external-link">Rsamtools</a></span><span class="op">)</span>  <span class="co"># for the BamFile() constructor function</span></span></span>
<span class="r-in"><span><span class="va">bamfile1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/BamFile-class.html" class="external-link">BamFile</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/pasillaBamSubset/man/pasillaBamSubset-package.html" class="external-link">untreated1_chr4</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">bamfile2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/BamFile-class.html" class="external-link">BamFile</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/pasillaBamSubset/man/pasillaBamSubset-package.html" class="external-link">untreated3_chr4</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">gpos5</span> <span class="op">&lt;-</span> <span class="fu">GPos</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqinfo</a></span><span class="op">(</span><span class="va">bamfile1</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicAlignments" class="external-link">GenomicAlignments</a></span><span class="op">)</span>  <span class="co"># for "coverage" method for BamFile objects</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: SummarizedExperiment</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: MatrixGenerics</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: matrixStats</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘matrixStats’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:Biobase’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     anyMissing, rowMedians</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘MatrixGenerics’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:matrixStats’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colWeightedMeans, colWeightedMedians, colWeightedSds,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowWeightedSds, rowWeightedVars</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following object is masked from ‘package:Biobase’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowMedians</span>
<span class="r-in"><span><span class="va">cvg1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="coverage-methods.html">coverage</a></span><span class="op">(</span><span class="va">bamfile1</span><span class="op">)</span>, use.names<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">cvg2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="coverage-methods.html">coverage</a></span><span class="op">(</span><span class="va">bamfile2</span><span class="op">)</span>, use.names<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">mcols</a></span><span class="op">(</span><span class="va">gpos5</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html" class="external-link">DataFrame</a></span><span class="op">(</span><span class="va">cvg1</span>, <span class="va">cvg2</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">gpos5</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> StitchedGPos object with 120748101 positions and 2 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>               seqnames       pos strand |  cvg1  cvg2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                  &lt;Rle&gt; &lt;integer&gt;  &lt;Rle&gt; | &lt;Rle&gt; &lt;Rle&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           [1]    chr2L         1      * |     0     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           [2]    chr2L         2      * |     0     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           [3]    chr2L         3      * |     0     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           [4]    chr2L         4      * |     0     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           [5]    chr2L         5      * |     0     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           ...      ...       ...    ... .   ...   ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [120748097]  chrYHet    347034      * |     0     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [120748098]  chrYHet    347035      * |     0     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [120748099]  chrYHet    347036      * |     0     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [120748100]  chrYHet    347037      * |     0     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [120748101]  chrYHet    347038      * |     0     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 8 sequences from an unspecified genome</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/object.size.html" class="external-link">object.size</a></span><span class="op">(</span><span class="va">gpos5</span><span class="op">)</span>  <span class="co"># lightweight</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1828208 bytes</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Keep only the positions where coverage is at least 10 in one of the</span></span></span>
<span class="r-in"><span><span class="co">## 2 samples:</span></span></span>
<span class="r-in"><span><span class="va">gpos5</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">mcols</a></span><span class="op">(</span><span class="va">gpos5</span><span class="op">)</span><span class="op">$</span><span class="va">cvg1</span> <span class="op">&gt;=</span> <span class="fl">10</span> <span class="op">|</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">mcols</a></span><span class="op">(</span><span class="va">gpos5</span><span class="op">)</span><span class="op">$</span><span class="va">cvg2</span> <span class="op">&gt;=</span> <span class="fl">10</span><span class="op">]</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> StitchedGPos object with 214773 positions and 2 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>            seqnames       pos strand |  cvg1  cvg2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>               &lt;Rle&gt; &lt;integer&gt;  &lt;Rle&gt; | &lt;Rle&gt; &lt;Rle&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        [1]     chr4      4936      * |    11     6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        [2]     chr4      4937      * |    12     6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        [3]     chr4      4938      * |    12     6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        [4]     chr4      4939      * |    13     6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        [5]     chr4      4940      * |    13     6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        ...      ...       ...    ... .   ...   ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [214769]     chr4   1348339      * |    10    14</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [214770]     chr4   1348340      * |     6    12</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [214771]     chr4   1348341      * |     6    10</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [214772]     chr4   1348342      * |     6    10</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [214773]     chr4   1348343      * |     0    10</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 8 sequences from an unspecified genome</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## ---------------------------------------------------------------------</span></span></span>
<span class="r-in"><span><span class="co">## USING A GPos OBJECT IN A RangedSummarizedExperiment OBJECT</span></span></span>
<span class="r-in"><span><span class="co">## ---------------------------------------------------------------------</span></span></span>
<span class="r-in"><span><span class="co">## Because the GPos class extends the GenomicRanges virtual class, a</span></span></span>
<span class="r-in"><span><span class="co">## GPos derivative can be used as the rowRanges component of a</span></span></span>
<span class="r-in"><span><span class="co">## RangedSummarizedExperiment object.</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## As a 1st example, we show how the counts for samples sA and sB in</span></span></span>
<span class="r-in"><span><span class="co">## 'gpos3' can be stored in a SummarizedExperiment object where the rows</span></span></span>
<span class="r-in"><span><span class="co">## correspond to unique genomic positions and the columns to samples:</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">SummarizedExperiment</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html" class="external-link">cbind</a></span><span class="op">(</span>sA<span class="op">=</span><span class="va">sA_counts</span>, sB<span class="op">=</span><span class="va">sB_counts</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">mcols</a></span><span class="op">(</span><span class="va">gpos3A</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span></span>
<span class="r-in"><span><span class="va">rse3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html" class="external-link">SummarizedExperiment</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>counts<span class="op">=</span><span class="va">counts</span><span class="op">)</span>, rowRanges<span class="op">=</span><span class="va">gpos3A</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">rse3</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> class: RangedSummarizedExperiment </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> dim: 1996 2 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> metadata(0):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> assays(1): counts</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rownames: NULL</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rowData names(0):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colnames(2): sA sB</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colData names(0):</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/pkg/MatrixGenerics/man/rowRanges.html" class="external-link">rowRanges</a></span><span class="op">(</span><span class="va">rse3</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> StitchedGPos object with 1996 positions and 0 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          seqnames       pos strand</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             &lt;Rle&gt; &lt;integer&gt;  &lt;Rle&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      [1]     chrI         1      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      [2]     chrI         2      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      [3]     chrI         3      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      [4]     chrI         4      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      [5]     chrI         5      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      ...      ...       ...    ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1992]     chrI      1996      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1993]     chrI      1997      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1994]     chrI      1998      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1995]     chrI      1999      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1996]     chrI      2000      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">rse3</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      sA sB</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1,]  9  9</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2,]  9  4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [3,] 10  3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [4,] 10  1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [5,]  7  7</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [6,]  2  2</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Finally we show how the coverage data from Example 5 can be easily</span></span></span>
<span class="r-in"><span><span class="co">## stored in a lightweight SummarizedExperiment derivative:</span></span></span>
<span class="r-in"><span><span class="va">cvg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">mcols</a></span><span class="op">(</span><span class="va">gpos5</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">mcols</a></span><span class="op">(</span><span class="va">gpos5</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span></span>
<span class="r-in"><span><span class="va">rse5</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html" class="external-link">SummarizedExperiment</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>cvg<span class="op">=</span><span class="va">cvg</span><span class="op">)</span>, rowRanges<span class="op">=</span><span class="va">gpos5</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">rse5</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> class: RangedSummarizedExperiment </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> dim: 120748101 2 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> metadata(0):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> assays(1): cvg</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rownames: NULL</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rowData names(0):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colnames(2): cvg1 cvg2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colData names(0):</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/pkg/MatrixGenerics/man/rowRanges.html" class="external-link">rowRanges</a></span><span class="op">(</span><span class="va">rse5</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> StitchedGPos object with 120748101 positions and 0 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>               seqnames       pos strand</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                  &lt;Rle&gt; &lt;integer&gt;  &lt;Rle&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           [1]    chr2L         1      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           [2]    chr2L         2      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           [3]    chr2L         3      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           [4]    chr2L         4      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           [5]    chr2L         5      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           ...      ...       ...    ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [120748097]  chrYHet    347034      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [120748098]  chrYHet    347035      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [120748099]  chrYHet    347036      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [120748100]  chrYHet    347037      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [120748101]  chrYHet    347038      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 8 sequences from an unspecified genome</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">rse5</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DataFrame with 120748101 rows and 2 columns</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>            cvg1  cvg2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           &lt;Rle&gt; &lt;Rle&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1             0     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2             0     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3             0     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4             0     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5             0     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ...         ...   ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 120748097     0     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 120748098     0     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 120748099     0     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 120748100     0     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 120748101     0     0</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Keep only the positions where coverage is at least 10 in one of the</span></span></span>
<span class="r-in"><span><span class="co">## 2 samples:</span></span></span>
<span class="r-in"><span><span class="va">rse5</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">rse5</span><span class="op">)</span><span class="op">$</span><span class="va">cvg1</span> <span class="op">&gt;=</span> <span class="fl">10</span> <span class="op">|</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">rse5</span><span class="op">)</span><span class="op">$</span><span class="va">cvg2</span> <span class="op">&gt;=</span> <span class="fl">10</span><span class="op">]</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> class: RangedSummarizedExperiment </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> dim: 214773 2 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> metadata(0):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> assays(1): cvg</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rownames: NULL</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rowData names(0):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colnames(2): cvg1 cvg2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colData names(0):</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Patrick Aboyoun, Hervé Pagès, Michael Lawrence.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

