<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>An Introduction to the GenomicRanges Package • GenomicRanges</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="An Introduction to the GenomicRanges Package">
<meta property="og:description" content="GenomicRanges">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GenomicRanges</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.51.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/GenomicRangesIntroduction.html">An Introduction to the GenomicRanges Package</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/Bioconductor/GenomicRanges/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>An Introduction to the GenomicRanges
Package</h1>
                        <h4 data-toc-skip class="author">Marc Carlson,
Patrick Aboyoun, Hervé Pagès, and Martin Morgan</h4>
            
            <h4 data-toc-skip class="date">December 19, 2022</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/Bioconductor/GenomicRanges/blob/HEAD/vignettes/GenomicRangesIntroduction.Rmd" class="external-link"><code>vignettes/GenomicRangesIntroduction.Rmd</code></a></small>
      <div class="hidden name"><code>GenomicRangesIntroduction.Rmd</code></div>

    </div>

    
    
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script><script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script><div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The <em><a href="https://bioconductor.org/packages/3.17/GenomicRanges" class="external-link">GenomicRanges</a></em>
package serves as the foundation for representing genomic locations
within the Bioconductor project. In the Bioconductor package hierarchy,
it builds upon the <em><a href="https://bioconductor.org/packages/3.17/IRanges" class="external-link">IRanges</a></em>
(infrastructure) package and provides support for the <em><a href="https://bioconductor.org/packages/3.17/BSgenome" class="external-link">BSgenome</a></em>
(infrastructure), <em><a href="https://bioconductor.org/packages/3.17/Rsamtools" class="external-link">Rsamtools</a></em>
(I/O), <em><a href="https://bioconductor.org/packages/3.17/ShortRead" class="external-link">ShortRead</a></em>
(I/O &amp; QA), <em><a href="https://bioconductor.org/packages/3.17/rtracklayer" class="external-link">rtracklayer</a></em>
(I/O), <em><a href="https://bioconductor.org/packages/3.17/GenomicFeatures" class="external-link">GenomicFeatures</a></em>
(infrastructure), <em><a href="https://bioconductor.org/packages/3.17/GenomicAlignments" class="external-link">GenomicAlignments</a></em>
(sequence reads), <em><a href="https://bioconductor.org/packages/3.17/VariantAnnotation" class="external-link">VariantAnnotation</a></em>
(called variants), and many other Bioconductor packages.</p>
<p>This package lays a foundation for genomic analysis by introducing
three classes (<em>GRanges</em>, <em>GPos</em>, and
<em>GRangesList</em>), which are used to represent genomic ranges,
genomic positions, and groups of genomic ranges. This vignette focuses
on the <em>GRanges</em> and <em>GRangesList</em> classes and their
associated methods.</p>
<p>The <em><a href="https://bioconductor.org/packages/3.17/GenomicRanges" class="external-link">GenomicRanges</a></em>
package is available at <a href="https://bioconductor.org" class="external-link">https://bioconductor.org</a> and can be
installed via <code><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">BiocManager::install</a></code>:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"GenomicRanges"</span><span class="op">)</span></span></code></pre></div>
<p>A package only needs to be installed once. Load the package into an R
session with</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges" class="external-link">GenomicRanges</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="granges-genomic-ranges">
<em>GRanges</em>: Genomic Ranges<a class="anchor" aria-label="anchor" href="#granges-genomic-ranges"></a>
</h2>
<p>The <em>GRanges</em> class represents a collection of genomic ranges
that each have a single start and end location on the genome. It can be
used to store the location of genomic features such as contiguous
binding sites, transcripts, and exons. These objects can be created by
using the <code>GRanges</code> constructor function. For example,</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span></span>
<span>    seqnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Rle-class.html" class="external-link">Rle</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chr1"</span>, <span class="st">"chr2"</span>, <span class="st">"chr1"</span>, <span class="st">"chr3"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span>, <span class="fl">2</span>, <span class="fl">4</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    ranges <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span><span class="fl">101</span><span class="op">:</span><span class="fl">110</span>, end <span class="op">=</span> <span class="fl">111</span><span class="op">:</span><span class="fl">120</span>, names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">letters</span>, <span class="fl">10</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    strand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Rle-class.html" class="external-link">Rle</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/strand.html" class="external-link">strand</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"-"</span>, <span class="st">"+"</span>, <span class="st">"*"</span>, <span class="st">"+"</span>, <span class="st">"-"</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">3</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    score <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,</span>
<span>    GC <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0</span>, length<span class="op">=</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">gr</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 10 ranges and 2 metadata columns:</span></span>
<span><span class="co">##     seqnames    ranges strand |     score        GC</span></span>
<span><span class="co">##        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   a     chr1   101-111      - |         1  1.000000</span></span>
<span><span class="co">##   b     chr2   102-112      + |         2  0.888889</span></span>
<span><span class="co">##   c     chr2   103-113      + |         3  0.777778</span></span>
<span><span class="co">##   d     chr2   104-114      * |         4  0.666667</span></span>
<span><span class="co">##   e     chr1   105-115      * |         5  0.555556</span></span>
<span><span class="co">##   f     chr1   106-116      + |         6  0.444444</span></span>
<span><span class="co">##   g     chr3   107-117      + |         7  0.333333</span></span>
<span><span class="co">##   h     chr3   108-118      + |         8  0.222222</span></span>
<span><span class="co">##   i     chr3   109-119      - |         9  0.111111</span></span>
<span><span class="co">##   j     chr3   110-120      - |        10  0.000000</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 3 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
<p>creates a <em>GRanges</em> object with 10 genomic ranges. The output
of the <em>GRanges</em> <code>show</code> method separates the
information into a left and right hand region that are separated by
<code>|</code> symbols. The genomic coordinates (seqnames, ranges, and
strand) are located on the left-hand side and the metadata columns
(annotation) are located on the right. For this example, the metadata is
comprised of <code>score</code> and <code>GC</code> information, but
almost anything can be stored in the metadata portion of a
<em>GRanges</em> object.</p>
<p>The components of the genomic coordinates within a <em>GRanges</em>
object can be extracted using the <code>seqnames</code>,
<code>ranges</code>, and <code>strand</code> accessor functions.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqnames</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## factor-Rle of length 10 with 4 runs</span></span>
<span><span class="co">##   Lengths:    1    3    2    4</span></span>
<span><span class="co">##   Values : chr1 chr2 chr1 chr3</span></span>
<span><span class="co">## Levels(3): chr1 chr2 chr3</span></span></code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/range-squeezers.html" class="external-link">ranges</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## IRanges object with 10 ranges and 0 metadata columns:</span></span>
<span><span class="co">##         start       end     width</span></span>
<span><span class="co">##     &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></span>
<span><span class="co">##   a       101       111        11</span></span>
<span><span class="co">##   b       102       112        11</span></span>
<span><span class="co">##   c       103       113        11</span></span>
<span><span class="co">##   d       104       114        11</span></span>
<span><span class="co">##   e       105       115        11</span></span>
<span><span class="co">##   f       106       116        11</span></span>
<span><span class="co">##   g       107       117        11</span></span>
<span><span class="co">##   h       108       118        11</span></span>
<span><span class="co">##   i       109       119        11</span></span>
<span><span class="co">##   j       110       120        11</span></span></code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/strand.html" class="external-link">strand</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## factor-Rle of length 10 with 5 runs</span></span>
<span><span class="co">##   Lengths: 1 2 2 3 2</span></span>
<span><span class="co">##   Values : - + * + -</span></span>
<span><span class="co">## Levels(3): + - *</span></span></code></pre>
<p>The genomic ranges can be extracted without corresponding metadata
with <code>granges</code></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/genomic-range-squeezers.html">granges</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 10 ranges and 0 metadata columns:</span></span>
<span><span class="co">##     seqnames    ranges strand</span></span>
<span><span class="co">##        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">##   a     chr1   101-111      -</span></span>
<span><span class="co">##   b     chr2   102-112      +</span></span>
<span><span class="co">##   c     chr2   103-113      +</span></span>
<span><span class="co">##   d     chr2   104-114      *</span></span>
<span><span class="co">##   e     chr1   105-115      *</span></span>
<span><span class="co">##   f     chr1   106-116      +</span></span>
<span><span class="co">##   g     chr3   107-117      +</span></span>
<span><span class="co">##   h     chr3   108-118      +</span></span>
<span><span class="co">##   i     chr3   109-119      -</span></span>
<span><span class="co">##   j     chr3   110-120      -</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 3 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
<p>Annotations for these coordinates can be extracted as a
<em>DataFrame</em> object using the <code>mcols</code> accessor.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">mcols</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 10 rows and 2 columns</span></span>
<span><span class="co">##       score        GC</span></span>
<span><span class="co">##   &lt;integer&gt; &lt;numeric&gt;</span></span>
<span><span class="co">## a         1  1.000000</span></span>
<span><span class="co">## b         2  0.888889</span></span>
<span><span class="co">## c         3  0.777778</span></span>
<span><span class="co">## d         4  0.666667</span></span>
<span><span class="co">## e         5  0.555556</span></span>
<span><span class="co">## f         6  0.444444</span></span>
<span><span class="co">## g         7  0.333333</span></span>
<span><span class="co">## h         8  0.222222</span></span>
<span><span class="co">## i         9  0.111111</span></span>
<span><span class="co">## j        10  0.000000</span></span></code></pre>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">mcols</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span><span class="op">$</span><span class="va">score</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1]  1  2  3  4  5  6  7  8  9 10</span></span></code></pre>
<p>Information about the lengths of the various sequences that the
ranges are aligned to can also be stored in the <em>GRanges</em> object.
So if this is data from <em>Homo sapiens</em>, we can set the values
as:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlengths</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">249250621</span>, <span class="fl">243199373</span>, <span class="fl">198022430</span><span class="op">)</span></span></code></pre></div>
<p>And then retrieves as:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlengths</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##      chr1      chr2      chr3 </span></span>
<span><span class="co">## 249250621 243199373 198022430</span></span></code></pre>
<p>Methods for accessing the <code>length</code> and <code>names</code>
have also been defined.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "a" "b" "c" "d" "e" "f" "g" "h" "i" "j"</span></span></code></pre>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 10</span></span></code></pre>
<div class="section level3">
<h3 id="splitting-and-combining-granges-objects">Splitting and combining <em>GRanges</em> objects<a class="anchor" aria-label="anchor" href="#splitting-and-combining-granges-objects"></a>
</h3>
<p><em>GRanges</em> objects can be devided into groups using the
<code>split</code> method. This produces a <em>GRangesList</em> object,
a class that will be discussed in detail in the next section.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/splitAsList.html" class="external-link">split</a></span><span class="op">(</span><span class="va">gr</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span>, each<span class="op">=</span><span class="fl">5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sp</span></span></code></pre></div>
<pre><code><span><span class="co">## GRangesList object of length 2:</span></span>
<span><span class="co">## $`1`</span></span>
<span><span class="co">## GRanges object with 5 ranges and 2 metadata columns:</span></span>
<span><span class="co">##     seqnames    ranges strand |     score        GC</span></span>
<span><span class="co">##        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   a     chr1   101-111      - |         1  1.000000</span></span>
<span><span class="co">##   b     chr2   102-112      + |         2  0.888889</span></span>
<span><span class="co">##   c     chr2   103-113      + |         3  0.777778</span></span>
<span><span class="co">##   d     chr2   104-114      * |         4  0.666667</span></span>
<span><span class="co">##   e     chr1   105-115      * |         5  0.555556</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 3 sequences from an unspecified genome</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $`2`</span></span>
<span><span class="co">## GRanges object with 5 ranges and 2 metadata columns:</span></span>
<span><span class="co">##     seqnames    ranges strand |     score        GC</span></span>
<span><span class="co">##        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   f     chr1   106-116      + |         6  0.444444</span></span>
<span><span class="co">##   g     chr3   107-117      + |         7  0.333333</span></span>
<span><span class="co">##   h     chr3   108-118      + |         8  0.222222</span></span>
<span><span class="co">##   i     chr3   109-119      - |         9  0.111111</span></span>
<span><span class="co">##   j     chr3   110-120      - |        10  0.000000</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 3 sequences from an unspecified genome</span></span></code></pre>
<p>Separate <em>GRanges</em> instances can be concatenated by using the
<code>c</code> and <code>append</code> methods.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">sp</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="va">sp</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 10 ranges and 2 metadata columns:</span></span>
<span><span class="co">##     seqnames    ranges strand |     score        GC</span></span>
<span><span class="co">##        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   a     chr1   101-111      - |         1  1.000000</span></span>
<span><span class="co">##   b     chr2   102-112      + |         2  0.888889</span></span>
<span><span class="co">##   c     chr2   103-113      + |         3  0.777778</span></span>
<span><span class="co">##   d     chr2   104-114      * |         4  0.666667</span></span>
<span><span class="co">##   e     chr1   105-115      * |         5  0.555556</span></span>
<span><span class="co">##   f     chr1   106-116      + |         6  0.444444</span></span>
<span><span class="co">##   g     chr3   107-117      + |         7  0.333333</span></span>
<span><span class="co">##   h     chr3   108-118      + |         8  0.222222</span></span>
<span><span class="co">##   i     chr3   109-119      - |         9  0.111111</span></span>
<span><span class="co">##   j     chr3   110-120      - |        10  0.000000</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 3 sequences from an unspecified genome</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="subsetting-granges-objects">Subsetting <em>GRanges</em> objects<a class="anchor" aria-label="anchor" href="#subsetting-granges-objects"></a>
</h3>
<p><em>GRanges</em> objects act like vectors of ranges, with the
expected vector-like subsetting operations available</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gr</span><span class="op">[</span><span class="fl">2</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 2 ranges and 2 metadata columns:</span></span>
<span><span class="co">##     seqnames    ranges strand |     score        GC</span></span>
<span><span class="co">##        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   b     chr2   102-112      + |         2  0.888889</span></span>
<span><span class="co">##   c     chr2   103-113      + |         3  0.777778</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 3 sequences from an unspecified genome</span></span></code></pre>
<p>A second argument to the <code>[</code> subset operator can be used
to specify which metadata columns to extract from the <em>GRanges</em>
object. For example,</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gr</span><span class="op">[</span><span class="fl">2</span><span class="op">:</span><span class="fl">3</span>, <span class="st">"GC"</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 2 ranges and 1 metadata column:</span></span>
<span><span class="co">##     seqnames    ranges strand |        GC</span></span>
<span><span class="co">##        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;</span></span>
<span><span class="co">##   b     chr2   102-112      + |  0.888889</span></span>
<span><span class="co">##   c     chr2   103-113      + |  0.777778</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 3 sequences from an unspecified genome</span></span></code></pre>
<p>Elements can also be assigned to the <em>GRanges</em> object. Here is
an example where the second row of a <em>GRanges</em> object is replaced
with the first row of <code>gr</code>.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">singles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/splitAsList.html" class="external-link">split</a></span><span class="op">(</span><span class="va">gr</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">grMod</span> <span class="op">&lt;-</span> <span class="va">gr</span></span>
<span><span class="va">grMod</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">singles</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">grMod</span>, n<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 3 ranges and 2 metadata columns:</span></span>
<span><span class="co">##     seqnames    ranges strand |     score        GC</span></span>
<span><span class="co">##        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   a     chr1   101-111      - |         1  1.000000</span></span>
<span><span class="co">##   b     chr1   101-111      - |         1  1.000000</span></span>
<span><span class="co">##   c     chr2   103-113      + |         3  0.777778</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 3 sequences from an unspecified genome</span></span></code></pre>
<p>There are methods to repeat, reverse, or select specific portions of
<em>GRanges</em> objects.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="va">singles</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, times <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 3 ranges and 2 metadata columns:</span></span>
<span><span class="co">##     seqnames    ranges strand |     score        GC</span></span>
<span><span class="co">##        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   b     chr2   102-112      + |         2  0.888889</span></span>
<span><span class="co">##   b     chr2   102-112      + |         2  0.888889</span></span>
<span><span class="co">##   b     chr2   102-112      + |         2  0.888889</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 3 sequences from an unspecified genome</span></span></code></pre>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">rev</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 10 ranges and 2 metadata columns:</span></span>
<span><span class="co">##     seqnames    ranges strand |     score        GC</span></span>
<span><span class="co">##        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   j     chr3   110-120      - |        10  0.000000</span></span>
<span><span class="co">##   i     chr3   109-119      - |         9  0.111111</span></span>
<span><span class="co">##   h     chr3   108-118      + |         8  0.222222</span></span>
<span><span class="co">##   g     chr3   107-117      + |         7  0.333333</span></span>
<span><span class="co">##   f     chr1   106-116      + |         6  0.444444</span></span>
<span><span class="co">##   e     chr1   105-115      * |         5  0.555556</span></span>
<span><span class="co">##   d     chr2   104-114      * |         4  0.666667</span></span>
<span><span class="co">##   c     chr2   103-113      + |         3  0.777778</span></span>
<span><span class="co">##   b     chr2   102-112      + |         2  0.888889</span></span>
<span><span class="co">##   a     chr1   101-111      - |         1  1.000000</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 3 sequences from an unspecified genome</span></span></code></pre>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gr</span>,n<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 2 ranges and 2 metadata columns:</span></span>
<span><span class="co">##     seqnames    ranges strand |     score        GC</span></span>
<span><span class="co">##        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   a     chr1   101-111      - |         1  1.000000</span></span>
<span><span class="co">##   b     chr2   102-112      + |         2  0.888889</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 3 sequences from an unspecified genome</span></span></code></pre>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">tail</a></span><span class="op">(</span><span class="va">gr</span>,n<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 2 ranges and 2 metadata columns:</span></span>
<span><span class="co">##     seqnames    ranges strand |     score        GC</span></span>
<span><span class="co">##        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   i     chr3   109-119      - |         9  0.111111</span></span>
<span><span class="co">##   j     chr3   110-120      - |        10  0.000000</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 3 sequences from an unspecified genome</span></span></code></pre>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/window.html" class="external-link">window</a></span><span class="op">(</span><span class="va">gr</span>, start<span class="op">=</span><span class="fl">2</span>,end<span class="op">=</span><span class="fl">4</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 3 ranges and 2 metadata columns:</span></span>
<span><span class="co">##     seqnames    ranges strand |     score        GC</span></span>
<span><span class="co">##        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   b     chr2   102-112      + |         2  0.888889</span></span>
<span><span class="co">##   c     chr2   103-113      + |         3  0.777778</span></span>
<span><span class="co">##   d     chr2   104-114      * |         4  0.666667</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 3 sequences from an unspecified genome</span></span></code></pre>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gr</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span>start<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">7</span><span class="op">)</span>, end<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>,<span class="fl">9</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 5 ranges and 2 metadata columns:</span></span>
<span><span class="co">##     seqnames    ranges strand |     score        GC</span></span>
<span><span class="co">##        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   b     chr2   102-112      + |         2  0.888889</span></span>
<span><span class="co">##   c     chr2   103-113      + |         3  0.777778</span></span>
<span><span class="co">##   g     chr3   107-117      + |         7  0.333333</span></span>
<span><span class="co">##   h     chr3   108-118      + |         8  0.222222</span></span>
<span><span class="co">##   i     chr3   109-119      - |         9  0.111111</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 3 sequences from an unspecified genome</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="basic-interval-operations-for-granges-objects">Basic interval operations for <em>GRanges</em> objects<a class="anchor" aria-label="anchor" href="#basic-interval-operations-for-granges-objects"></a>
</h3>
<p>Basic interval characteristics of <em>GRanges</em> objects can be
extracted using the <code>start</code>, <code>end</code>,
<code>width</code>, and <code>range</code> methods.</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">g</span> <span class="op">&lt;-</span> <span class="va">gr</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span>
<span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/append.html" class="external-link">append</a></span><span class="op">(</span><span class="va">g</span>, <span class="va">singles</span><span class="op">[[</span><span class="fl">10</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/start.html" class="external-link">start</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 101 102 103 110</span></span></code></pre>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/start.html" class="external-link">end</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 111 112 113 120</span></span></code></pre>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/start.html" class="external-link">width</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 11 11 11 11</span></span></code></pre>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/inter-range-methods.html">range</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 3 ranges and 0 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">##   [1]     chr1   101-111      -</span></span>
<span><span class="co">##   [2]     chr2   102-113      +</span></span>
<span><span class="co">##   [3]     chr3   110-120      -</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 3 sequences from an unspecified genome</span></span></code></pre>
<p>The <em>GRanges</em> class also has many methods for manipulating the
ranges. The methods can be classified as <em>intra-range methods</em>,
<em>inter-range methods</em>, and <em>between-range methods</em>.</p>
<p><em>Intra-range methods</em> operate on each element of a
<em>GRanges</em> object independent of the other ranges in the object.
For example, the <code>flank</code> method can be used to recover
regions flanking the set of ranges represented by the <em>GRanges</em>
object. So to get a <em>GRanges</em> object containing the ranges that
include the 10 bases upstream of the ranges:</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/intra-range-methods.html">flank</a></span><span class="op">(</span><span class="va">g</span>, <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 4 ranges and 2 metadata columns:</span></span>
<span><span class="co">##     seqnames    ranges strand |     score        GC</span></span>
<span><span class="co">##        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   a     chr1   112-121      - |         1  1.000000</span></span>
<span><span class="co">##   b     chr2    92-101      + |         2  0.888889</span></span>
<span><span class="co">##   c     chr2    93-102      + |         3  0.777778</span></span>
<span><span class="co">##   j     chr3   121-130      - |        10  0.000000</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 3 sequences from an unspecified genome</span></span></code></pre>
<p>And to include the downstream bases:</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/intra-range-methods.html">flank</a></span><span class="op">(</span><span class="va">g</span>, <span class="fl">10</span>, start<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 4 ranges and 2 metadata columns:</span></span>
<span><span class="co">##     seqnames    ranges strand |     score        GC</span></span>
<span><span class="co">##        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   a     chr1    91-100      - |         1  1.000000</span></span>
<span><span class="co">##   b     chr2   113-122      + |         2  0.888889</span></span>
<span><span class="co">##   c     chr2   114-123      + |         3  0.777778</span></span>
<span><span class="co">##   j     chr3   100-109      - |        10  0.000000</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 3 sequences from an unspecified genome</span></span></code></pre>
<p>Other examples of intra-range methods include <code>resize</code> and
<code>shift</code>. The <code>shift</code> method will move the ranges
by a specific number of base pairs, and the <code>resize</code> method
will extend the ranges by a specified width.</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/intra-range-methods.html" class="external-link">shift</a></span><span class="op">(</span><span class="va">g</span>, <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 4 ranges and 2 metadata columns:</span></span>
<span><span class="co">##     seqnames    ranges strand |     score        GC</span></span>
<span><span class="co">##        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   a     chr1   106-116      - |         1  1.000000</span></span>
<span><span class="co">##   b     chr2   107-117      + |         2  0.888889</span></span>
<span><span class="co">##   c     chr2   108-118      + |         3  0.777778</span></span>
<span><span class="co">##   j     chr3   115-125      - |        10  0.000000</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 3 sequences from an unspecified genome</span></span></code></pre>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/intra-range-methods.html">resize</a></span><span class="op">(</span><span class="va">g</span>, <span class="fl">30</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 4 ranges and 2 metadata columns:</span></span>
<span><span class="co">##     seqnames    ranges strand |     score        GC</span></span>
<span><span class="co">##        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   a     chr1    82-111      - |         1  1.000000</span></span>
<span><span class="co">##   b     chr2   102-131      + |         2  0.888889</span></span>
<span><span class="co">##   c     chr2   103-132      + |         3  0.777778</span></span>
<span><span class="co">##   j     chr3    91-120      - |        10  0.000000</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 3 sequences from an unspecified genome</span></span></code></pre>
<p>The <em><a href="https://bioconductor.org/packages/3.17/GenomicRanges" class="external-link">GenomicRanges</a></em>
help page <code><a href="../reference/intra-range-methods.html">?"intra-range-methods"</a></code> summarizes these
methods.</p>
<p><em>Inter-range methods</em> involve comparisons between ranges in a
single <em>GRanges</em> object. For instance, the <code>reduce</code>
method will align the ranges and merge overlapping ranges to produce a
simplified set.</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/inter-range-methods.html">reduce</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 3 ranges and 0 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">##   [1]     chr1   101-111      -</span></span>
<span><span class="co">##   [2]     chr2   102-113      +</span></span>
<span><span class="co">##   [3]     chr3   110-120      -</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 3 sequences from an unspecified genome</span></span></code></pre>
<p>Sometimes one is interested in the gaps or the qualities of the gaps
between the ranges represented by your <em>GRanges</em> object. The
<code>gaps</code> method provides this information: reduced version of
your ranges:</p>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/inter-range-methods.html">gaps</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 12 ranges and 0 metadata columns:</span></span>
<span><span class="co">##        seqnames        ranges strand</span></span>
<span><span class="co">##           &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">##    [1]     chr1   1-249250621      +</span></span>
<span><span class="co">##    [2]     chr1         1-100      -</span></span>
<span><span class="co">##    [3]     chr1 112-249250621      -</span></span>
<span><span class="co">##    [4]     chr1   1-249250621      *</span></span>
<span><span class="co">##    [5]     chr2         1-101      +</span></span>
<span><span class="co">##    ...      ...           ...    ...</span></span>
<span><span class="co">##    [8]     chr2   1-243199373      *</span></span>
<span><span class="co">##    [9]     chr3   1-198022430      +</span></span>
<span><span class="co">##   [10]     chr3         1-109      -</span></span>
<span><span class="co">##   [11]     chr3 121-198022430      -</span></span>
<span><span class="co">##   [12]     chr3   1-198022430      *</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 3 sequences from an unspecified genome</span></span></code></pre>
<p>The <code>disjoin</code> method represents a <em>GRanges</em> object
as a collection of non-overlapping ranges:</p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/inter-range-methods.html">disjoin</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 5 ranges and 0 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">##   [1]     chr1   101-111      -</span></span>
<span><span class="co">##   [2]     chr2       102      +</span></span>
<span><span class="co">##   [3]     chr2   103-112      +</span></span>
<span><span class="co">##   [4]     chr2       113      +</span></span>
<span><span class="co">##   [5]     chr3   110-120      -</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 3 sequences from an unspecified genome</span></span></code></pre>
<p>The <code>coverage</code> method quantifies the degree of overlap for
all the ranges in a <em>GRanges</em> object.</p>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/coverage-methods.html">coverage</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## RleList of length 3</span></span>
<span><span class="co">## $chr1</span></span>
<span><span class="co">## integer-Rle of length 249250621 with 3 runs</span></span>
<span><span class="co">##   Lengths:       100        11 249250510</span></span>
<span><span class="co">##   Values :         0         1         0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $chr2</span></span>
<span><span class="co">## integer-Rle of length 243199373 with 5 runs</span></span>
<span><span class="co">##   Lengths:       101         1        10         1 243199260</span></span>
<span><span class="co">##   Values :         0         1         2         1         0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $chr3</span></span>
<span><span class="co">## integer-Rle of length 198022430 with 3 runs</span></span>
<span><span class="co">##   Lengths:       109        11 198022310</span></span>
<span><span class="co">##   Values :         0         1         0</span></span></code></pre>
<p>See the <em><a href="https://bioconductor.org/packages/3.17/GenomicRanges" class="external-link">GenomicRanges</a></em>
help page <code><a href="../reference/inter-range-methods.html">?"inter-range-methods"</a></code> for additional help.</p>
<p><em>Between-range methods</em> involve operations between two
<em>GRanges</em> objects; some of these are summarized in the next
section.</p>
</div>
<div class="section level3">
<h3 id="interval-set-operations-for-granges-objects">Interval set operations for <em>GRanges</em> objects<a class="anchor" aria-label="anchor" href="#interval-set-operations-for-granges-objects"></a>
</h3>
<p><em>Between-range methods</em> calculate relationships between
different <em>GRanges</em> objects. Of central importance are
<code>findOverlaps</code> and related operations; these are discussed
below. Additional operations treat <em>GRanges</em> as mathematical sets
of coordinates; <code>union(g, g2)</code> is the union of the
coordinates in <code>g</code> and <code>g2</code>. Here are examples for
calculating the <code>union</code>, the <code>intersect</code> and the
asymmetric difference (using <code>setdiff</code>).</p>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">g2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gr</span>, n<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/setops-methods.html">union</a></span><span class="op">(</span><span class="va">g</span>, <span class="va">g2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 3 ranges and 0 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">##   [1]     chr1   101-111      -</span></span>
<span><span class="co">##   [2]     chr2   102-113      +</span></span>
<span><span class="co">##   [3]     chr3   110-120      -</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 3 sequences from an unspecified genome</span></span></code></pre>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/setops-methods.html">intersect</a></span><span class="op">(</span><span class="va">g</span>, <span class="va">g2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 2 ranges and 0 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">##   [1]     chr1   101-111      -</span></span>
<span><span class="co">##   [2]     chr2   102-112      +</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 3 sequences from an unspecified genome</span></span></code></pre>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/setops-methods.html">setdiff</a></span><span class="op">(</span><span class="va">g</span>, <span class="va">g2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 2 ranges and 0 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">##   [1]     chr2       113      +</span></span>
<span><span class="co">##   [2]     chr3   110-120      -</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 3 sequences from an unspecified genome</span></span></code></pre>
<p>Related methods are available when the structure of the
<em>GRanges</em> objects are ‘parallel’ to one another, i.e., element 1
of object 1 is related to element 1 of object 2, and so on. These
operations all begin with a <code>p</code>, which is short for parallel.
The methods then perform element-wise, e.g., the union of element 1 of
object 1 with element 1 of object 2, etc. A requirement for these
operations is that the number of elements in each <em>GRanges</em>
object is the same, and that both of the objects have the same seqnames
and strand assignments throughout.</p>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">g3</span> <span class="op">&lt;-</span> <span class="va">g</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/range-squeezers.html" class="external-link">ranges</a></span><span class="op">(</span><span class="va">g3</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span>start<span class="op">=</span><span class="fl">105</span>, end<span class="op">=</span><span class="fl">112</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/setops-methods.html">punion</a></span><span class="op">(</span><span class="va">g2</span>, <span class="va">g3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 2 ranges and 0 metadata columns:</span></span>
<span><span class="co">##     seqnames    ranges strand</span></span>
<span><span class="co">##        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">##   a     chr1   101-112      -</span></span>
<span><span class="co">##   b     chr2   102-112      +</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 3 sequences from an unspecified genome</span></span></code></pre>
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/setops-methods.html">pintersect</a></span><span class="op">(</span><span class="va">g2</span>, <span class="va">g3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 2 ranges and 3 metadata columns:</span></span>
<span><span class="co">##     seqnames    ranges strand |     score        GC       hit</span></span>
<span><span class="co">##        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt; &lt;logical&gt;</span></span>
<span><span class="co">##   a     chr1   105-111      - |         1  1.000000      TRUE</span></span>
<span><span class="co">##   b     chr2   102-112      + |         2  0.888889      TRUE</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 3 sequences from an unspecified genome</span></span></code></pre>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/setops-methods.html">psetdiff</a></span><span class="op">(</span><span class="va">g2</span>, <span class="va">g3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 2 ranges and 0 metadata columns:</span></span>
<span><span class="co">##     seqnames    ranges strand</span></span>
<span><span class="co">##        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">##   a     chr1   101-104      -</span></span>
<span><span class="co">##   b     chr2   102-101      +</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 3 sequences from an unspecified genome</span></span></code></pre>
<p>For more information on the <code>GRanges</code> class be sure to
consult the manual page.</p>
<div class="sourceCode" id="cb82"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">?</span><span class="va">GRanges</span></span></code></pre></div>
<p>A relatively comprehensive list of available methods is discovered
with</p>
<div class="sourceCode" id="cb83"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/methods.html" class="external-link">methods</a></span><span class="op">(</span>class<span class="op">=</span><span class="st">"GRanges"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="grangeslist-groups-of-genomic-ranges">
<em>GRangesList</em>: Groups of Genomic Ranges<a class="anchor" aria-label="anchor" href="#grangeslist-groups-of-genomic-ranges"></a>
</h2>
<p>Some important genomic features, such as spliced transcripts that are
comprised of exons, are inherently compound structures. Such a feature
makes much more sense when expressed as a compound object such as a
<em>GRangesList</em>. Whenever genomic features consist of multiple
ranges that are grouped by a parent feature, they can be represented as
a <em>GRangesList</em> object. Consider the simple example of the two
transcript <code>GRangesList</code> below created using the
<code>GRangesList</code> constructor.</p>
<div class="sourceCode" id="cb84"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gr1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span></span>
<span>    seqnames <span class="op">=</span> <span class="st">"chr2"</span>,</span>
<span>    ranges <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span><span class="fl">103</span>, <span class="fl">106</span><span class="op">)</span>,</span>
<span>    strand <span class="op">=</span> <span class="st">"+"</span>,</span>
<span>    score <span class="op">=</span> <span class="fl">5L</span>, GC <span class="op">=</span> <span class="fl">0.45</span><span class="op">)</span></span>
<span><span class="va">gr2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span></span>
<span>    seqnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chr1"</span>, <span class="st">"chr1"</span><span class="op">)</span>,</span>
<span>    ranges <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">107</span>, <span class="fl">113</span><span class="op">)</span>, width <span class="op">=</span> <span class="fl">3</span><span class="op">)</span>,</span>
<span>    strand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+"</span>, <span class="st">"-"</span><span class="op">)</span>,</span>
<span>    score <span class="op">=</span> <span class="fl">3</span><span class="op">:</span><span class="fl">4</span>, GC <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.3</span>, <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">grl</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GRangesList-class.html">GRangesList</a></span><span class="op">(</span><span class="st">"txA"</span> <span class="op">=</span> <span class="va">gr1</span>, <span class="st">"txB"</span> <span class="op">=</span> <span class="va">gr2</span><span class="op">)</span></span>
<span><span class="va">grl</span></span></code></pre></div>
<pre><code><span><span class="co">## GRangesList object of length 2:</span></span>
<span><span class="co">## $txA</span></span>
<span><span class="co">## GRanges object with 1 range and 2 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand |     score        GC</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   [1]     chr2   103-106      + |         5      0.45</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $txB</span></span>
<span><span class="co">## GRanges object with 2 ranges and 2 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand |     score        GC</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   [1]     chr1   107-109      + |         3       0.3</span></span>
<span><span class="co">##   [2]     chr1   113-115      - |         4       0.5</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
<p>The <code>show</code> method for a <em>GRangesList</em> object
displays it as a named list of <em>GRanges</em> objects, where the names
of this list are considered to be the names of the grouping feature. In
the example above, the groups of individual exon ranges are represented
as separate <em>GRanges</em> objects which are further organized into a
list structure where each element name is a transcript name. Many other
combinations of grouped and labeled <em>GRanges</em> objects are
possible of course, but this example is expected to be a common
arrangement.</p>
<div class="section level3">
<h3 id="basic-grangeslist-accessors">Basic <em>GRangesList</em> accessors<a class="anchor" aria-label="anchor" href="#basic-grangeslist-accessors"></a>
</h3>
<p>Just as with <em>GRanges</em> object, the components of the genomic
coordinates within a <em>GRangesList</em> object can be extracted using
simple accessor methods. Not surprisingly, the <em>GRangesList</em>
objects have many of the same accessors as <em>GRanges</em> objects. The
difference is that many of these methods return a list since the input
is now essentially a list of <em>GRanges</em> objects. Here are a few
examples:</p>
<div class="sourceCode" id="cb86"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqnames</a></span><span class="op">(</span><span class="va">grl</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## RleList of length 2</span></span>
<span><span class="co">## $txA</span></span>
<span><span class="co">## factor-Rle of length 1 with 1 run</span></span>
<span><span class="co">##   Lengths:    1</span></span>
<span><span class="co">##   Values : chr2</span></span>
<span><span class="co">## Levels(2): chr2 chr1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $txB</span></span>
<span><span class="co">## factor-Rle of length 2 with 1 run</span></span>
<span><span class="co">##   Lengths:    2</span></span>
<span><span class="co">##   Values : chr1</span></span>
<span><span class="co">## Levels(2): chr2 chr1</span></span></code></pre>
<div class="sourceCode" id="cb88"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/range-squeezers.html" class="external-link">ranges</a></span><span class="op">(</span><span class="va">grl</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## IRangesList object of length 2:</span></span>
<span><span class="co">## $txA</span></span>
<span><span class="co">## IRanges object with 1 range and 0 metadata columns:</span></span>
<span><span class="co">##           start       end     width</span></span>
<span><span class="co">##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></span>
<span><span class="co">##   [1]       103       106         4</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $txB</span></span>
<span><span class="co">## IRanges object with 2 ranges and 0 metadata columns:</span></span>
<span><span class="co">##           start       end     width</span></span>
<span><span class="co">##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></span>
<span><span class="co">##   [1]       107       109         3</span></span>
<span><span class="co">##   [2]       113       115         3</span></span></code></pre>
<div class="sourceCode" id="cb90"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/strand.html" class="external-link">strand</a></span><span class="op">(</span><span class="va">grl</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## RleList of length 2</span></span>
<span><span class="co">## $txA</span></span>
<span><span class="co">## factor-Rle of length 1 with 1 run</span></span>
<span><span class="co">##   Lengths: 1</span></span>
<span><span class="co">##   Values : +</span></span>
<span><span class="co">## Levels(3): + - *</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $txB</span></span>
<span><span class="co">## factor-Rle of length 2 with 2 runs</span></span>
<span><span class="co">##   Lengths: 1 1</span></span>
<span><span class="co">##   Values : + -</span></span>
<span><span class="co">## Levels(3): + - *</span></span></code></pre>
<p>The <code>length</code> and <code>names</code> methods will return
the length or names of the list and the <code>seqlengths</code> method
will return the set of sequence lengths.</p>
<div class="sourceCode" id="cb92"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">grl</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 2</span></span></code></pre>
<div class="sourceCode" id="cb94"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">grl</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "txA" "txB"</span></span></code></pre>
<div class="sourceCode" id="cb96"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlengths</a></span><span class="op">(</span><span class="va">grl</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## chr2 chr1 </span></span>
<span><span class="co">##   NA   NA</span></span></code></pre>
<p>The <code>elementNROWS</code> method returns a list of integers
corresponding to the result of calling <code>NROW</code> on each
individual <em>GRanges</em> object contained by the
<em>GRangesList</em>. This is a faster alternative to calling
<code>lapply</code> on the <em>GRangesList</em>.</p>
<div class="sourceCode" id="cb98"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/List-class.html" class="external-link">elementNROWS</a></span><span class="op">(</span><span class="va">grl</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## txA txB </span></span>
<span><span class="co">##   1   2</span></span></code></pre>
<p><code>isEmpty</code> tests if a <em>GRangesList</em> object contains
anything.</p>
<div class="sourceCode" id="cb100"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/List-class.html" class="external-link">isEmpty</a></span><span class="op">(</span><span class="va">grl</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] FALSE</span></span></code></pre>
<p>In the context of a <em>GRangesList</em> object, the
<code>mcols</code> method performs a similar operation to what it does
on a <em>GRanges</em> object. However, this metadata now refers to
information at the list level instead of the level of the individual
<em>GRanges</em> objects.</p>
<div class="sourceCode" id="cb102"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">mcols</a></span><span class="op">(</span><span class="va">grl</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Transcript A"</span>,<span class="st">"Transcript B"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">mcols</a></span><span class="op">(</span><span class="va">grl</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 2 rows and 1 column</span></span>
<span><span class="co">##            value</span></span>
<span><span class="co">##      &lt;character&gt;</span></span>
<span><span class="co">## txA Transcript A</span></span>
<span><span class="co">## txB Transcript B</span></span></code></pre>
<p>Element-level metadata can be retrieved by unlisting the
<code>GRangesList</code>, and extracting the metadata</p>
<div class="sourceCode" id="cb104"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">mcols</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">grl</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 3 rows and 2 columns</span></span>
<span><span class="co">##         score        GC</span></span>
<span><span class="co">##     &lt;integer&gt; &lt;numeric&gt;</span></span>
<span><span class="co">## txA         5      0.45</span></span>
<span><span class="co">## txB         3      0.30</span></span>
<span><span class="co">## txB         4      0.50</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="combining-grangeslist-objects">Combining <em>GRangesList</em> objects<a class="anchor" aria-label="anchor" href="#combining-grangeslist-objects"></a>
</h3>
<p><em>GRangesList</em> objects can be unlisted to combine the separate
<em>GRanges</em> objects that they contain as an expanded
<em>GRanges</em>.</p>
<div class="sourceCode" id="cb106"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ul</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">grl</span><span class="op">)</span></span>
<span><span class="va">ul</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 3 ranges and 2 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand |     score        GC</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   txA     chr2   103-106      + |         5      0.45</span></span>
<span><span class="co">##   txB     chr1   107-109      + |         3      0.30</span></span>
<span><span class="co">##   txB     chr1   113-115      - |         4      0.50</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
<p>Append lists using <code>append</code> or <code>c</code>.</p>
<p>A <a href="https://support.bioconductor.org/p/89339/" class="external-link">support site
user</a> had two <em>GRangesList</em> objects with ‘parallel’ elements,
and wanted to combined these element-wise into a single
<em>GRangesList</em>. One solution is to use <code><a href="https://rdrr.io/pkg/S4Vectors/man/List-utils.html" class="external-link">pc()</a></code> –
parallel (element-wise) <code><a href="https://rdrr.io/r/base/c.html" class="external-link">c()</a></code>. A more general solution is to
concatenate the lists and then re-group by some factor, in this case the
names of the elements.</p>
<div class="sourceCode" id="cb108"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">grl1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GRangesList-class.html">GRangesList</a></span><span class="op">(</span></span>
<span>    gr1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="st">"chr2"</span>, <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">6</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    gr2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="st">"chr1"</span>, <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">7</span>,<span class="fl">13</span><span class="op">)</span>, width <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">grl2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GRangesList-class.html">GRangesList</a></span><span class="op">(</span></span>
<span>    gr1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="st">"chr2"</span>, <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span><span class="fl">9</span>, <span class="fl">12</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    gr2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="st">"chr1"</span>, <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">25</span>,<span class="fl">38</span><span class="op">)</span>, width <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/List-utils.html" class="external-link">pc</a></span><span class="op">(</span><span class="va">grl1</span>, <span class="va">grl2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRangesList object of length 2:</span></span>
<span><span class="co">## $gr1</span></span>
<span><span class="co">## GRanges object with 2 ranges and 0 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">##   [1]     chr2       3-6      *</span></span>
<span><span class="co">##   [2]     chr2      9-12      *</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $gr2</span></span>
<span><span class="co">## GRanges object with 4 ranges and 0 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">##   [1]     chr1       7-9      *</span></span>
<span><span class="co">##   [2]     chr1     13-15      *</span></span>
<span><span class="co">##   [3]     chr1     25-27      *</span></span>
<span><span class="co">##   [4]     chr1     38-40      *</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
<div class="sourceCode" id="cb110"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">grl3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">grl1</span>, <span class="va">grl2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/extractList.html" class="external-link">regroup</a></span><span class="op">(</span><span class="va">grl3</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">grl3</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRangesList object of length 2:</span></span>
<span><span class="co">## $gr1</span></span>
<span><span class="co">## GRanges object with 2 ranges and 0 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">##   [1]     chr2       3-6      *</span></span>
<span><span class="co">##   [2]     chr2      9-12      *</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $gr2</span></span>
<span><span class="co">## GRanges object with 4 ranges and 0 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">##   [1]     chr1       7-9      *</span></span>
<span><span class="co">##   [2]     chr1     13-15      *</span></span>
<span><span class="co">##   [3]     chr1     25-27      *</span></span>
<span><span class="co">##   [4]     chr1     38-40      *</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="basic-interval-operations-for-grangeslist-objects">Basic interval operations for <em>GRangesList</em> objects<a class="anchor" aria-label="anchor" href="#basic-interval-operations-for-grangeslist-objects"></a>
</h3>
<p>For interval operations, many of the same methods exist for
<em>GRangesList</em> objects that exist for <em>GRanges</em>
objects.</p>
<div class="sourceCode" id="cb112"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/start.html" class="external-link">start</a></span><span class="op">(</span><span class="va">grl</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## IntegerList of length 2</span></span>
<span><span class="co">## [["txA"]] 103</span></span>
<span><span class="co">## [["txB"]] 107 113</span></span></code></pre>
<div class="sourceCode" id="cb114"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/start.html" class="external-link">end</a></span><span class="op">(</span><span class="va">grl</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## IntegerList of length 2</span></span>
<span><span class="co">## [["txA"]] 106</span></span>
<span><span class="co">## [["txB"]] 109 115</span></span></code></pre>
<div class="sourceCode" id="cb116"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/start.html" class="external-link">width</a></span><span class="op">(</span><span class="va">grl</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## IntegerList of length 2</span></span>
<span><span class="co">## [["txA"]] 4</span></span>
<span><span class="co">## [["txB"]] 3 3</span></span></code></pre>
<p>These operations return a data structure representing, e.g.,
<em>IntegerList</em>, a list where all elements are integers; it can be
convenient to use mathematical and other operations on <em>List</em>
objects that work on each element, e.g.,</p>
<div class="sourceCode" id="cb118"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/start.html" class="external-link">width</a></span><span class="op">(</span><span class="va">grl</span><span class="op">)</span><span class="op">)</span>  <span class="co"># sum of widths of each grl element</span></span></code></pre></div>
<pre><code><span><span class="co">## txA txB </span></span>
<span><span class="co">##   4   6</span></span></code></pre>
<p>Most of the intra-, inter- and between-range methods operate on
<em>GRangesList</em> objects, e.g., to shift all the <em>GRanges</em>
objects in a <em>GRangesList</em> object, or calculate the coverage.
Both of these operations are also carried out across each
<em>GRanges</em> list member.</p>
<div class="sourceCode" id="cb120"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/intra-range-methods.html" class="external-link">shift</a></span><span class="op">(</span><span class="va">grl</span>, <span class="fl">20</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRangesList object of length 2:</span></span>
<span><span class="co">## $txA</span></span>
<span><span class="co">## GRanges object with 1 range and 2 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand |     score        GC</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   [1]     chr2   123-126      + |         5      0.45</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $txB</span></span>
<span><span class="co">## GRanges object with 2 ranges and 2 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand |     score        GC</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   [1]     chr1   127-129      + |         3       0.3</span></span>
<span><span class="co">##   [2]     chr1   133-135      - |         4       0.5</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
<div class="sourceCode" id="cb122"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/coverage-methods.html">coverage</a></span><span class="op">(</span><span class="va">grl</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## RleList of length 2</span></span>
<span><span class="co">## $chr2</span></span>
<span><span class="co">## integer-Rle of length 106 with 2 runs</span></span>
<span><span class="co">##   Lengths: 102   4</span></span>
<span><span class="co">##   Values :   0   1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $chr1</span></span>
<span><span class="co">## integer-Rle of length 115 with 4 runs</span></span>
<span><span class="co">##   Lengths: 106   3   3   3</span></span>
<span><span class="co">##   Values :   0   1   0   1</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="subsetting-grangeslist-objects">Subsetting <em>GRangesList</em> objects<a class="anchor" aria-label="anchor" href="#subsetting-grangeslist-objects"></a>
</h3>
<p>A <em>GRangesList</em> object behaves like a <code>list</code>:
<code>[</code> returns a <em>GRangesList</em> containing a subset of the
original object; <code>[[</code> or <code>$</code> returns the
<em>GRanges</em> object at that location in the list.</p>
<div class="sourceCode" id="cb124"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">grl</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">grl</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">grl</span><span class="op">[</span><span class="st">"txA"</span><span class="op">]</span></span>
<span><span class="va">grl</span><span class="op">$</span><span class="va">txB</span></span></code></pre></div>
<p>In addition, subsetting a <em>GRangesList</em> also accepts a second
parameter to specify which of the metadata columns you wish to
select.</p>
<div class="sourceCode" id="cb125"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">grl</span><span class="op">[</span><span class="fl">1</span>, <span class="st">"score"</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## GRangesList object of length 1:</span></span>
<span><span class="co">## $txA</span></span>
<span><span class="co">## GRanges object with 1 range and 1 metadata column:</span></span>
<span><span class="co">##       seqnames    ranges strand |     score</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;</span></span>
<span><span class="co">##   [1]     chr2   103-106      + |         5</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
<div class="sourceCode" id="cb127"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">grl</span><span class="op">[</span><span class="st">"txB"</span>, <span class="st">"GC"</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## GRangesList object of length 1:</span></span>
<span><span class="co">## $txB</span></span>
<span><span class="co">## GRanges object with 2 ranges and 1 metadata column:</span></span>
<span><span class="co">##       seqnames    ranges strand |        GC</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;</span></span>
<span><span class="co">##   [1]     chr1   107-109      + |       0.3</span></span>
<span><span class="co">##   [2]     chr1   113-115      - |       0.5</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
<p>The <code>head</code>, <code>tail</code>, <code>rep</code>,
<code>rev</code>, and <code>window</code> methods all behave as you
would expect them to for a list object. For example, the elements
referred to by <code>window</code> are now list elements instead of
<em>GRanges</em> elements.</p>
<div class="sourceCode" id="cb129"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="va">grl</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, times <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 3 ranges and 2 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand |     score        GC</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   [1]     chr2   103-106      + |         5      0.45</span></span>
<span><span class="co">##   [2]     chr2   103-106      + |         5      0.45</span></span>
<span><span class="co">##   [3]     chr2   103-106      + |         5      0.45</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
<div class="sourceCode" id="cb131"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">rev</a></span><span class="op">(</span><span class="va">grl</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRangesList object of length 2:</span></span>
<span><span class="co">## $txB</span></span>
<span><span class="co">## GRanges object with 2 ranges and 2 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand |     score        GC</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   [1]     chr1   107-109      + |         3       0.3</span></span>
<span><span class="co">##   [2]     chr1   113-115      - |         4       0.5</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $txA</span></span>
<span><span class="co">## GRanges object with 1 range and 2 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand |     score        GC</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   [1]     chr2   103-106      + |         5      0.45</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
<div class="sourceCode" id="cb133"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">grl</span>, n<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRangesList object of length 1:</span></span>
<span><span class="co">## $txA</span></span>
<span><span class="co">## GRanges object with 1 range and 2 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand |     score        GC</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   [1]     chr2   103-106      + |         5      0.45</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
<div class="sourceCode" id="cb135"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">tail</a></span><span class="op">(</span><span class="va">grl</span>, n<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRangesList object of length 1:</span></span>
<span><span class="co">## $txB</span></span>
<span><span class="co">## GRanges object with 2 ranges and 2 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand |     score        GC</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   [1]     chr1   107-109      + |         3       0.3</span></span>
<span><span class="co">##   [2]     chr1   113-115      - |         4       0.5</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
<div class="sourceCode" id="cb137"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/window.html" class="external-link">window</a></span><span class="op">(</span><span class="va">grl</span>, start<span class="op">=</span><span class="fl">1</span>, end<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRangesList object of length 1:</span></span>
<span><span class="co">## $txA</span></span>
<span><span class="co">## GRanges object with 1 range and 2 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand |     score        GC</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   [1]     chr2   103-106      + |         5      0.45</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
<div class="sourceCode" id="cb139"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">grl</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span>start<span class="op">=</span><span class="fl">2</span>, end<span class="op">=</span><span class="fl">2</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## GRangesList object of length 1:</span></span>
<span><span class="co">## $txB</span></span>
<span><span class="co">## GRanges object with 2 ranges and 2 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand |     score        GC</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   [1]     chr1   107-109      + |         3       0.3</span></span>
<span><span class="co">##   [2]     chr1   113-115      - |         4       0.5</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="looping-over-grangeslist-objects">Looping over <em>GRangesList</em> objects<a class="anchor" aria-label="anchor" href="#looping-over-grangeslist-objects"></a>
</h3>
<p>For <em>GRangesList</em> objects there is also a family of
<code>apply</code> methods. These include <code>lapply</code>,
<code>sapply</code>, <code>mapply</code>, <code>endoapply</code>,
<code>mendoapply</code>, <code>Map</code>, and <code>Reduce</code>.</p>
<p>The different looping methods defined for <em>GRangesList</em>
objects are useful for returning different kinds of results. The
standard <code>lapply</code> and <code>sapply</code> behave according to
convention, with the <code>lapply</code> method returning a list and
<code>sapply</code> returning a more simplified output.</p>
<div class="sourceCode" id="cb141"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">grl</span>, <span class="va">length</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $txA</span></span>
<span><span class="co">## [1] 1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $txB</span></span>
<span><span class="co">## [1] 2</span></span></code></pre>
<div class="sourceCode" id="cb143"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">grl</span>, <span class="va">length</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## txA txB </span></span>
<span><span class="co">##   1   2</span></span></code></pre>
<p>As with <em>IRanges</em> objects, there is also a multivariate
version of <code>sapply</code>, called <code>mapply</code>, defined for
<em>GRangesList</em> objects. And, if you don’t want the results
simplified, you can call the <code>Map</code> method, which does the
same things as <code>mapply</code> but without simplifying the
output.</p>
<div class="sourceCode" id="cb145"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">grl2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/intra-range-methods.html" class="external-link">shift</a></span><span class="op">(</span><span class="va">grl</span>, <span class="fl">10</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">grl2</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"shiftTxA"</span>, <span class="st">"shiftTxB"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="va">c</span>, <span class="va">grl</span>, <span class="va">grl2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $txA</span></span>
<span><span class="co">## GRanges object with 2 ranges and 2 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand |     score        GC</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   [1]     chr2   103-106      + |         5      0.45</span></span>
<span><span class="co">##   [2]     chr2   113-116      + |         5      0.45</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $txB</span></span>
<span><span class="co">## GRanges object with 4 ranges and 2 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand |     score        GC</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   [1]     chr1   107-109      + |         3       0.3</span></span>
<span><span class="co">##   [2]     chr1   113-115      - |         4       0.5</span></span>
<span><span class="co">##   [3]     chr1   117-119      + |         3       0.3</span></span>
<span><span class="co">##   [4]     chr1   123-125      - |         4       0.5</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
<div class="sourceCode" id="cb147"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/funprog.html" class="external-link">Map</a></span><span class="op">(</span><span class="va">c</span>, <span class="va">grl</span>, <span class="va">grl2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $txA</span></span>
<span><span class="co">## GRanges object with 2 ranges and 2 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand |     score        GC</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   [1]     chr2   103-106      + |         5      0.45</span></span>
<span><span class="co">##   [2]     chr2   113-116      + |         5      0.45</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $txB</span></span>
<span><span class="co">## GRanges object with 4 ranges and 2 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand |     score        GC</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   [1]     chr1   107-109      + |         3       0.3</span></span>
<span><span class="co">##   [2]     chr1   113-115      - |         4       0.5</span></span>
<span><span class="co">##   [3]     chr1   117-119      + |         3       0.3</span></span>
<span><span class="co">##   [4]     chr1   123-125      - |         4       0.5</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
<p>Sometimes you will want to get back a modified version of the
<em>GRangesList</em> that you originally passed in.</p>
<p>An endomorphism is a transformation of an object to another instance
of the same class . This is achieved using the <code>endoapply</code>
method, which will return the results as a <em>GRangesList</em>
object.</p>
<div class="sourceCode" id="cb149"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/List-utils.html" class="external-link">endoapply</a></span><span class="op">(</span><span class="va">grl</span>, <span class="va">rev</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRangesList object of length 2:</span></span>
<span><span class="co">## $txA</span></span>
<span><span class="co">## GRanges object with 1 range and 2 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand |     score        GC</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   [1]     chr2   103-106      + |         5      0.45</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $txB</span></span>
<span><span class="co">## GRanges object with 2 ranges and 2 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand |     score        GC</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   [1]     chr1   113-115      - |         4       0.5</span></span>
<span><span class="co">##   [2]     chr1   107-109      + |         3       0.3</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
<div class="sourceCode" id="cb151"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/List-utils.html" class="external-link">mendoapply</a></span><span class="op">(</span><span class="va">c</span>, <span class="va">grl</span>, <span class="va">grl2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRangesList object of length 2:</span></span>
<span><span class="co">## $txA</span></span>
<span><span class="co">## GRanges object with 2 ranges and 2 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand |     score        GC</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   [1]     chr2   103-106      + |         5      0.45</span></span>
<span><span class="co">##   [2]     chr2   113-116      + |         5      0.45</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $txB</span></span>
<span><span class="co">## GRanges object with 4 ranges and 2 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand |     score        GC</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   [1]     chr1   107-109      + |         3       0.3</span></span>
<span><span class="co">##   [2]     chr1   113-115      - |         4       0.5</span></span>
<span><span class="co">##   [3]     chr1   117-119      + |         3       0.3</span></span>
<span><span class="co">##   [4]     chr1   123-125      - |         4       0.5</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
<p>The <code>Reduce</code> method will allow the <em>GRanges</em>
objects to be collapsed across the whole of the <em>GRangesList</em>
object. % Again, this seems like a sub-optimal example to me.</p>
<div class="sourceCode" id="cb153"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/funprog.html" class="external-link">Reduce</a></span><span class="op">(</span><span class="va">c</span>, <span class="va">grl</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 3 ranges and 2 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand |     score        GC</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   [1]     chr2   103-106      + |         5      0.45</span></span>
<span><span class="co">##   [2]     chr1   107-109      + |         3      0.30</span></span>
<span><span class="co">##   [3]     chr1   113-115      - |         4      0.50</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
<p>Explicit element-wise operations (<code><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply()</a></code> and friends)
on <em>GRangesList</em> objects with many elements can be slow. It is
therefore beneficial to explore operations that work on <em>List</em>
objects directly (e.g., many of the ‘group generic’ operators, see
<code><a href="https://rdrr.io/r/methods/S4groupGeneric.html" class="external-link">?S4groupGeneric</a></code>, and the set and parallel set operators
(e.g., <code>union</code>, <code>punion</code>). A useful and fast
strategy is to <code>unlist</code> the <em>GRangesList</em> to a
<em>GRanges</em> object, operate on the <em>GRanges</em> object, then
<code>relist</code> the result, e.g.,</p>
<div class="sourceCode" id="cb155"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">grl</span><span class="op">)</span></span>
<span><span class="va">gr</span><span class="op">$</span><span class="va">log_score</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">gr</span><span class="op">$</span><span class="va">score</span><span class="op">)</span></span>
<span><span class="va">grl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/relist.html" class="external-link">relist</a></span><span class="op">(</span><span class="va">gr</span>, <span class="va">grl</span><span class="op">)</span></span>
<span><span class="va">grl</span></span></code></pre></div>
<pre><code><span><span class="co">## GRangesList object of length 2:</span></span>
<span><span class="co">## $txA</span></span>
<span><span class="co">## GRanges object with 1 range and 3 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand |     score        GC log_score</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   txA     chr2   103-106      + |         5      0.45   1.60944</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $txB</span></span>
<span><span class="co">## GRanges object with 2 ranges and 3 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand |     score        GC log_score</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   txB     chr1   107-109      + |         3       0.3   1.09861</span></span>
<span><span class="co">##   txB     chr1   113-115      - |         4       0.5   1.38629</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
<p>See also <code><a href="https://rdrr.io/pkg/IRanges/man/extractList.html" class="external-link">?extractList</a></code>.</p>
<p>For more information on the <code>GRangesList</code> class be sure to
consult the manual page and available methods</p>
<div class="sourceCode" id="cb157"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">?</span><span class="va">GRangesList</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/methods.html" class="external-link">methods</a></span><span class="op">(</span>class<span class="op">=</span><span class="st">"GRangesList"</span><span class="op">)</span>   <span class="co"># _partial_ list</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="interval-overlaps-involving-granges-and-grangeslist-objects">Interval overlaps involving <em>GRanges</em> and
<em>GRangesList</em> objects<a class="anchor" aria-label="anchor" href="#interval-overlaps-involving-granges-and-grangeslist-objects"></a>
</h2>
<p>Interval overlapping is the process of comparing the ranges in two
objects to determine if and when they overlap. As such, it is perhaps
the most common operation performed on <em>GRanges</em> and
<em>GRangesList</em> objects. To this end, the <em><a href="https://bioconductor.org/packages/3.17/GenomicRanges" class="external-link">GenomicRanges</a></em>
package provides a family of interval overlap functions. The most
general of these functions is <code>findOverlaps</code>, which takes a
query and a subject as inputs and returns a <em>Hits</em> object
containing the index pairings for the overlapping elements.</p>
<div class="sourceCode" id="cb158"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/findOverlaps-methods.html">findOverlaps</a></span><span class="op">(</span><span class="va">gr</span>, <span class="va">grl</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Hits object with 3 hits and 0 metadata columns:</span></span>
<span><span class="co">##       queryHits subjectHits</span></span>
<span><span class="co">##       &lt;integer&gt;   &lt;integer&gt;</span></span>
<span><span class="co">##   [1]         1           1</span></span>
<span><span class="co">##   [2]         2           2</span></span>
<span><span class="co">##   [3]         3           2</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   queryLength: 3 / subjectLength: 2</span></span></code></pre>
<p>As suggested in the sections discussing the nature of the
<em>GRanges</em> and <em>GRangesList</em> classes, the index in the
above <em>Hits</em> object for a <em>GRanges</em> object is a single
range while for a <em>GRangesList</em> object it is the set of ranges
that define a “feature”.</p>
<p>Another function in the overlaps family is
<code>countOverlaps</code>, which tabulates the number of overlaps for
each element in the query.</p>
<div class="sourceCode" id="cb160"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/findOverlaps-methods.html">countOverlaps</a></span><span class="op">(</span><span class="va">gr</span>, <span class="va">grl</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## txA txB txB </span></span>
<span><span class="co">##   1   1   1</span></span></code></pre>
<p>A third function in this family is <code>subsetByOverlaps</code>,
which extracts the elements in the query that overlap at least one
element in the subject.</p>
<div class="sourceCode" id="cb162"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/findOverlaps-methods.html">subsetByOverlaps</a></span><span class="op">(</span><span class="va">gr</span>,<span class="va">grl</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 3 ranges and 3 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand |     score        GC log_score</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   txA     chr2   103-106      + |         5      0.45   1.60944</span></span>
<span><span class="co">##   txB     chr1   107-109      + |         3      0.30   1.09861</span></span>
<span><span class="co">##   txB     chr1   113-115      - |         4      0.50   1.38629</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
<p>Finally, you can use the <code>select</code> argument to get the
index of the first overlapping element in the subject for each element
in the query.</p>
<div class="sourceCode" id="cb164"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/findOverlaps-methods.html">findOverlaps</a></span><span class="op">(</span><span class="va">gr</span>, <span class="va">grl</span>, select<span class="op">=</span><span class="st">"first"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 1 2 2</span></span></code></pre>
<div class="sourceCode" id="cb166"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/findOverlaps-methods.html">findOverlaps</a></span><span class="op">(</span><span class="va">grl</span>, <span class="va">gr</span>, select<span class="op">=</span><span class="st">"first"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 1 2</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="finding-the-nearest-genomic-position-in-granges-objects">Finding the nearest genomic position in <em>GRanges</em>
objects<a class="anchor" aria-label="anchor" href="#finding-the-nearest-genomic-position-in-granges-objects"></a>
</h2>
<p>The <em><a href="https://bioconductor.org/packages/3.17/GenomicRanges" class="external-link">GenomicRanges</a></em>
package provides multiple functions to facilitate the indentification of
neighboring genomic positions. For the following examples, we define an
arbitrary <em>GRanges</em> object for <code>x</code> and we define the
<em>GRanges</em> object <code>subject</code> as the collection of genes
in <em><a href="https://bioconductor.org/packages/3.17/TxDb.Hsapiens.UCSC.hg19.knownGene" class="external-link">TxDb.Hsapiens.UCSC.hg19.knownGene</a></em>
extracted using the <code>genes</code> method from the <em><a href="https://bioconductor.org/packages/3.17/GenomicFeatures" class="external-link">GenomicFeatures</a></em>
package.</p>
<div class="sourceCode" id="cb168"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">txdb</span> <span class="op">&lt;-</span> <span class="fu">TxDb.Hsapiens.UCSC.hg19.knownGene</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/TxDb.Hsapiens.UCSC.hg19.knownGene/man/package.html" class="external-link">TxDb.Hsapiens.UCSC.hg19.knownGene</a></span></span>
<span><span class="va">broads</span> <span class="op">&lt;-</span> <span class="fu">GenomicFeatures</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/transcripts.html" class="external-link">genes</a></span><span class="op">(</span><span class="va">txdb</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   403 genes were dropped because they have exons located on both strands</span></span>
<span><span class="co">##   of the same reference sequence or on more than one reference sequence,</span></span>
<span><span class="co">##   so cannot be represented by a single genomic range.</span></span>
<span><span class="co">##   Use 'single.strand.genes.only=FALSE' to get all the genes in a</span></span>
<span><span class="co">##   GRangesList object, or use suppressMessages() to suppress this message.</span></span></code></pre>
<div class="sourceCode" id="cb170"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span></span>
<span>    seqnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Rle-class.html" class="external-link">Rle</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chr1"</span>, <span class="st">"chr2"</span>, <span class="st">"chr1"</span>, <span class="st">"chr3"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span>, <span class="fl">2</span>, <span class="fl">4</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    ranges <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span><span class="fl">101</span><span class="op">:</span><span class="fl">110</span>, end <span class="op">=</span> <span class="fl">111</span><span class="op">:</span><span class="fl">120</span>, names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">letters</span>, <span class="fl">10</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    strand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Rle-class.html" class="external-link">Rle</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/strand.html" class="external-link">strand</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"-"</span>, <span class="st">"+"</span>, <span class="st">"*"</span>, <span class="st">"+"</span>, <span class="st">"-"</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">3</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    score <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, GC <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0</span>, length<span class="op">=</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">subject</span> <span class="op">&lt;-</span> <span class="va">broads</span><span class="op">[</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqnames</a></span><span class="op">(</span><span class="va">broads</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlevels</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span> <span class="op">]</span></span></code></pre></div>
<p>The <code>nearest</code> method performs conventional nearest
neighbor finding. It finds the nearest neighbor range in
<code>subject</code> for each range in <code>x</code>. Overlaps are
included. If <code>subject</code> is not given as an argument,
<code>x</code> will also be treated as the <code>subject</code>.</p>
<div class="sourceCode" id="cb171"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/nearest-methods.html">nearest</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">subject</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] 2898 2224 2224 2765   73   73   NA   NA   NA   NA</span></span></code></pre>
<div class="sourceCode" id="cb173"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/nearest-methods.html">nearest</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1]  5  4  4  3  6  5  8  7 10  9</span></span></code></pre>
<p>The <code>precede</code> method will return the index of the range in
<code>subject</code> that is preceded by the range in <code>x</code>.
Overlaps are excluded.</p>
<div class="sourceCode" id="cb175"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/nearest-methods.html">precede</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">subject</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1]   NA 2224 2224 2224   73   73   NA   NA   NA   NA</span></span></code></pre>
<p>The <code>follow</code> method will return the index of the range in
<code>subject</code> that is followed by the range in
<code>x</code>.</p>
<div class="sourceCode" id="cb177"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/nearest-methods.html">follow</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">subject</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] 2898   NA   NA 2765 2898   NA   NA   NA   NA   NA</span></span></code></pre>
<p>The <code>nearestKNeighbors</code> method performs conventional
k-nearest neighbor finding. For each range in <code>x</code>, it will
find the index of the k-nearest neighbors in <code>subject</code>. The
argument <code>k</code> can be specified to identify more than one
nearest neighbor. Overlaps are included. If <code>subject</code> is not
given as an argument, <code>x</code> will also be treated as the
<code>subject</code>.</p>
<div class="sourceCode" id="cb179"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/nearest-methods.html">nearestKNeighbors</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">subject</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## IntegerList of length 10</span></span>
<span><span class="co">## [[1]] 2898</span></span>
<span><span class="co">## [[2]] 2224</span></span>
<span><span class="co">## [[3]] 2224</span></span>
<span><span class="co">## [[4]] 2765</span></span>
<span><span class="co">## [[5]] 73</span></span>
<span><span class="co">## [[6]] 73</span></span>
<span><span class="co">## [[7]] &lt;NA&gt;</span></span>
<span><span class="co">## [[8]] &lt;NA&gt;</span></span>
<span><span class="co">## [[9]] &lt;NA&gt;</span></span>
<span><span class="co">## [[10]] &lt;NA&gt;</span></span></code></pre>
<div class="sourceCode" id="cb181"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/nearest-methods.html">nearestKNeighbors</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">subject</span>, k<span class="op">=</span><span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## IntegerList of length 10</span></span>
<span><span class="co">## [[1]] 2898 3094 80 3240 1441 32 1312 3448 2609 1912</span></span>
<span><span class="co">## [[2]] 2224 2260 3040 3104 2149 177 2695 3165 2468 2926</span></span>
<span><span class="co">## [[3]] 2224 2260 3040 3104 2149 177 2695 3165 2468 2926</span></span>
<span><span class="co">## [[4]] 2765 2224 2260 3040 3104 2149 177 2695 3165 2468</span></span>
<span><span class="co">## [[5]] 73 2898 3191 2783 756 1606 3367 3722 1763 1234</span></span>
<span><span class="co">## [[6]] 73 3191 2783 756 1606 3367 3722 1763 1234 1941</span></span>
<span><span class="co">## [[7]] &lt;NA&gt;</span></span>
<span><span class="co">## [[8]] &lt;NA&gt;</span></span>
<span><span class="co">## [[9]] &lt;NA&gt;</span></span>
<span><span class="co">## [[10]] &lt;NA&gt;</span></span></code></pre>
<div class="sourceCode" id="cb183"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/nearest-methods.html">nearestKNeighbors</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## IntegerList of length 10</span></span>
<span><span class="co">## [[1]] 5</span></span>
<span><span class="co">## [[2]] 3</span></span>
<span><span class="co">## [[3]] 2</span></span>
<span><span class="co">## [[4]] 2</span></span>
<span><span class="co">## [[5]] 1</span></span>
<span><span class="co">## [[6]] 5</span></span>
<span><span class="co">## [[7]] 8</span></span>
<span><span class="co">## [[8]] 7</span></span>
<span><span class="co">## [[9]] 10</span></span>
<span><span class="co">## [[10]] 9</span></span></code></pre>
<div class="sourceCode" id="cb185"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/nearest-methods.html">nearestKNeighbors</a></span><span class="op">(</span><span class="va">x</span>, k<span class="op">=</span><span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## IntegerList of length 10</span></span>
<span><span class="co">## [[1]] 5 5</span></span>
<span><span class="co">## [[2]] 3 4</span></span>
<span><span class="co">## [[3]] 2 4</span></span>
<span><span class="co">## [[4]] 2 3</span></span>
<span><span class="co">## [[5]] 1 6 1</span></span>
<span><span class="co">## [[6]] 5</span></span>
<span><span class="co">## [[7]] 8</span></span>
<span><span class="co">## [[8]] 7</span></span>
<span><span class="co">## [[9]] 10 10</span></span>
<span><span class="co">## [[10]] 9 9</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="session-information">Session Information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<p>All of the output in this vignette was produced under the following
conditions:</p>
<div class="sourceCode" id="cb187"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R Under development (unstable) (2022-12-14 r83463)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.1 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] GenomicRanges_1.51.5 GenomeInfoDb_1.35.8  IRanges_2.33.0      </span></span>
<span><span class="co">## [4] S4Vectors_0.37.3     BiocGenerics_0.45.0  BiocStyle_2.27.0    </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] tidyselect_1.2.0                       </span></span>
<span><span class="co">##  [2] dplyr_1.0.10                           </span></span>
<span><span class="co">##  [3] blob_1.2.3                             </span></span>
<span><span class="co">##  [4] filelock_1.0.2                         </span></span>
<span><span class="co">##  [5] Biostrings_2.67.0                      </span></span>
<span><span class="co">##  [6] bitops_1.0-7                           </span></span>
<span><span class="co">##  [7] fastmap_1.1.0                          </span></span>
<span><span class="co">##  [8] RCurl_1.98-1.9                         </span></span>
<span><span class="co">##  [9] BiocFileCache_2.7.1                    </span></span>
<span><span class="co">## [10] GenomicAlignments_1.35.0               </span></span>
<span><span class="co">## [11] XML_3.99-0.13                          </span></span>
<span><span class="co">## [12] digest_0.6.31                          </span></span>
<span><span class="co">## [13] lifecycle_1.0.3                        </span></span>
<span><span class="co">## [14] ellipsis_0.3.2                         </span></span>
<span><span class="co">## [15] KEGGREST_1.39.0                        </span></span>
<span><span class="co">## [16] RSQLite_2.2.19                         </span></span>
<span><span class="co">## [17] magrittr_2.0.3                         </span></span>
<span><span class="co">## [18] compiler_4.3.0                         </span></span>
<span><span class="co">## [19] rlang_1.0.6                            </span></span>
<span><span class="co">## [20] sass_0.4.4                             </span></span>
<span><span class="co">## [21] progress_1.2.2                         </span></span>
<span><span class="co">## [22] tools_4.3.0                            </span></span>
<span><span class="co">## [23] utf8_1.2.2                             </span></span>
<span><span class="co">## [24] yaml_2.3.6                             </span></span>
<span><span class="co">## [25] rtracklayer_1.59.0                     </span></span>
<span><span class="co">## [26] knitr_1.41                             </span></span>
<span><span class="co">## [27] prettyunits_1.1.1                      </span></span>
<span><span class="co">## [28] bit_4.0.5                              </span></span>
<span><span class="co">## [29] curl_4.3.3                             </span></span>
<span><span class="co">## [30] DelayedArray_0.25.0                    </span></span>
<span><span class="co">## [31] xml2_1.3.3                             </span></span>
<span><span class="co">## [32] BiocParallel_1.33.7                    </span></span>
<span><span class="co">## [33] purrr_0.3.5                            </span></span>
<span><span class="co">## [34] desc_1.4.2                             </span></span>
<span><span class="co">## [35] grid_4.3.0                             </span></span>
<span><span class="co">## [36] fansi_1.0.3                            </span></span>
<span><span class="co">## [37] SummarizedExperiment_1.29.1            </span></span>
<span><span class="co">## [38] biomaRt_2.55.0                         </span></span>
<span><span class="co">## [39] cli_3.4.1                              </span></span>
<span><span class="co">## [40] rmarkdown_2.19                         </span></span>
<span><span class="co">## [41] crayon_1.5.2                           </span></span>
<span><span class="co">## [42] ragg_1.2.4                             </span></span>
<span><span class="co">## [43] generics_0.1.3                         </span></span>
<span><span class="co">## [44] httr_1.4.4                             </span></span>
<span><span class="co">## [45] rjson_0.2.21                           </span></span>
<span><span class="co">## [46] DBI_1.1.3                              </span></span>
<span><span class="co">## [47] cachem_1.0.6                           </span></span>
<span><span class="co">## [48] stringr_1.5.0                          </span></span>
<span><span class="co">## [49] zlibbioc_1.45.0                        </span></span>
<span><span class="co">## [50] assertthat_0.2.1                       </span></span>
<span><span class="co">## [51] parallel_4.3.0                         </span></span>
<span><span class="co">## [52] AnnotationDbi_1.61.0                   </span></span>
<span><span class="co">## [53] BiocManager_1.30.19                    </span></span>
<span><span class="co">## [54] XVector_0.39.0                         </span></span>
<span><span class="co">## [55] restfulr_0.0.15                        </span></span>
<span><span class="co">## [56] matrixStats_0.63.0                     </span></span>
<span><span class="co">## [57] vctrs_0.5.1                            </span></span>
<span><span class="co">## [58] Matrix_1.5-3                           </span></span>
<span><span class="co">## [59] jsonlite_1.8.4                         </span></span>
<span><span class="co">## [60] bookdown_0.31                          </span></span>
<span><span class="co">## [61] hms_1.1.2                              </span></span>
<span><span class="co">## [62] bit64_4.0.5                            </span></span>
<span><span class="co">## [63] systemfonts_1.0.4                      </span></span>
<span><span class="co">## [64] GenomicFeatures_1.51.2                 </span></span>
<span><span class="co">## [65] jquerylib_0.1.4                        </span></span>
<span><span class="co">## [66] glue_1.6.2                             </span></span>
<span><span class="co">## [67] pkgdown_2.0.7                          </span></span>
<span><span class="co">## [68] codetools_0.2-18                       </span></span>
<span><span class="co">## [69] stringi_1.7.8                          </span></span>
<span><span class="co">## [70] BiocIO_1.9.1                           </span></span>
<span><span class="co">## [71] tibble_3.1.8                           </span></span>
<span><span class="co">## [72] pillar_1.8.1                           </span></span>
<span><span class="co">## [73] rappdirs_0.3.3                         </span></span>
<span><span class="co">## [74] htmltools_0.5.4                        </span></span>
<span><span class="co">## [75] GenomeInfoDbData_1.2.9                 </span></span>
<span><span class="co">## [76] R6_2.5.1                               </span></span>
<span><span class="co">## [77] dbplyr_2.2.1                           </span></span>
<span><span class="co">## [78] textshaping_0.3.6                      </span></span>
<span><span class="co">## [79] rprojroot_2.0.3                        </span></span>
<span><span class="co">## [80] lattice_0.20-45                        </span></span>
<span><span class="co">## [81] evaluate_0.19                          </span></span>
<span><span class="co">## [82] Biobase_2.59.0                         </span></span>
<span><span class="co">## [83] png_0.1-8                              </span></span>
<span><span class="co">## [84] Rsamtools_2.15.0                       </span></span>
<span><span class="co">## [85] memoise_2.0.1                          </span></span>
<span><span class="co">## [86] TxDb.Hsapiens.UCSC.hg19.knownGene_3.2.2</span></span>
<span><span class="co">## [87] bslib_0.4.2                            </span></span>
<span><span class="co">## [88] Rcpp_1.0.9                             </span></span>
<span><span class="co">## [89] xfun_0.35                              </span></span>
<span><span class="co">## [90] MatrixGenerics_1.11.0                  </span></span>
<span><span class="co">## [91] fs_1.5.2                               </span></span>
<span><span class="co">## [92] pkgconfig_2.0.3</span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Patrick Aboyoun, Hervé Pagès, Michael Lawrence.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
